{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26a7e76d-96a8-487b-b2f4-933da56f2430",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-03-04 13:17:56.338942: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2023-03-04 13:17:56.453834: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/slurm/lib64:\n",
      "2023-03-04 13:17:56.453852: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2023-03-04 13:17:56.479805: E tensorflow/stream_executor/cuda/cuda_blas.cc:2981] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2023-03-04 13:17:57.285717: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/slurm/lib64:\n",
      "2023-03-04 13:17:57.285910: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/slurm/lib64:\n",
      "2023-03-04 13:17:57.285918: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib \n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "from sklearn import metrics\n",
    "import tabulate\n",
    "import scipy\n",
    "import itertools\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.lines as mlines\n",
    "from collections import Counter\n",
    "import umap\n",
    "from importlib import reload\n",
    "from scipy import stats\n",
    "import matplotlib.image as mpimg\n",
    "import goatools\n",
    "from goatools.anno.gaf_reader import GafReader\n",
    "import Bio.UniProt.GOA as GOA\n",
    "from sklearn import metrics\n",
    "from sklearn.decomposition import PCA\n",
    "import tabulate\n",
    "import itertools\n",
    "import igraph as ig\n",
    "import matplotlib.patches as mpatches\n",
    "from sklearn.neighbors import radius_neighbors_graph\n",
    "from scipy.ndimage import gaussian_filter1d\n",
    "from sklearn.neighbors import kneighbors_graph\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "import leidenalg as la\n",
    "import dask.array as da\n",
    "from sklearn.cluster import SpectralClustering\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "import umap\n",
    "import bbknn\n",
    "import bbknn.matrix\n",
    "from importlib import reload\n",
    "from scipy import stats\n",
    "import matplotlib.image as mpimg\n",
    "from goatools.anno.gaf_reader import GafReader\n",
    "import Bio.UniProt.GOA as GOA\n",
    "import gget\n",
    "import sys\n",
    "\n",
    "# locals\n",
    "sys.path.append(\"../\")\n",
    "import utils as ut\n",
    "reload(ut)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e77a594-f041-42b5-8720-e5455d5eaec2",
   "metadata": {},
   "outputs": [],
   "source": [
    "clusterPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/scanpy/clusters.csv\"\n",
    "cf = pd.read_parquet(clusterPath)\n",
    "print(cf.shape)\n",
    "print(cf['cellType'].unique())\n",
    "cf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45ba266f-aa31-463a-88c7-05a4c14dba22",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"load scRNA\"\"\"\n",
    "dirPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/scanpy/\"\n",
    "\n",
    "rna = {}\n",
    "\n",
    "for f in os.listdir(dirPath):\n",
    "    \n",
    "    if f.endswith('.h5ad'):\n",
    "        key = f.split(\".\")[0]\n",
    "        fullPath = f\"{dirPath}{f}\"\n",
    "        data = sc.read(fullPath)\n",
    "        rna[key] = data\n",
    "    \n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eaf019a-42ac-479a-89a6-335e3f530985",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot clusters using individual embeddings\n",
    "# shows how well defined the clusters are at a single time point\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 12, 4\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "ncolors = cf['cellType'].nunique()\n",
    "ctypes = sorted(cf['cellType'].unique())\n",
    "\n",
    "colordict = dict(zip(ctypes, ut.ncolor(ncolors, cmap='Spectral')))\n",
    "\n",
    "fig, axs = plt.subplots(1, 3, sharey=True)\n",
    "axs = axs.ravel()\n",
    "\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    data = rna[key]\n",
    "        \n",
    "    # get the run key\n",
    "    allKeys = list(data.obs.keys() )\n",
    "    allKeys.remove('n_genes')\n",
    "    runKey = allKeys[0]\n",
    "    \n",
    "    clusterIds = data.obs[runKey].astype(int) + 1  # note that these are zero-indexed!\n",
    "    embedding = data.obsm['X_umap']\n",
    "    \n",
    "    # get coloring\n",
    "    c = ['k'] * len(embedding)\n",
    "    handles = []\n",
    "    \n",
    "    for ctype in ctypes:\n",
    "        mask = (cf['cellType'] == ctype) & (cf['key'] == key)\n",
    "        cids = cf[mask]['clusterId'].unique()\n",
    "        c = np.where(clusterIds.isin(cids), colordict[ctype], c)\n",
    "        \n",
    "        handle = mpatches.Patch(color=colordict[ctype], \n",
    "                                ec='k', \n",
    "                                label=ctype)\n",
    "        handles.append(handle)\n",
    "        \n",
    "    \n",
    "    axs[i].scatter(embedding[:, 0] , \n",
    "                   embedding[:, 1] , \n",
    "                   c=c,\n",
    "                   edgecolor=\"k\",\n",
    "                   linewidth=0.25,\n",
    "                   s=20,\n",
    "                   alpha=0.8,)\n",
    "    \n",
    "        \n",
    "    axs[i].set_title(key)\n",
    "    axs[i].set_xlabel(\"UMAP 1\")\n",
    "    \n",
    "    if i == 0:\n",
    "        axs[i].set_ylabel(\"UMAP 2\")\n",
    "    \n",
    "    _ = axs[i].set_xticks([])\n",
    "    _ = axs[i].set_yticks([])\n",
    "\n",
    "    \n",
    "plt.legend(handles=handles, bbox_to_anchor=(1.04, 1.02))\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f64fd203-edb1-4a67-b710-de979887ae05",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 6, 6\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "plt.legend(handles=handles, \n",
    "           bbox_to_anchor=(2, 1.02),\n",
    "           ncols=7)\n",
    "plt.axis(False)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a684a547-2076-48c2-945e-b4289b054225",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bfb0184-c3eb-4bdf-9c48-46f656adb863",
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 12, 4\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "\n",
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "fig, axs = plt.subplots(1, 3, sharey=True)\n",
    "axs = axs.ravel()\n",
    "\n",
    "res = []\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    data = rna[key]\n",
    "    \n",
    "    clusterCount = pd.DataFrame(data.obs[runKey].astype(int).value_counts(normalize=True))\n",
    "    clusterCount.columns = ['proportion'] \n",
    "    clusterCount['clusterId'] = clusterCount.index + 1\n",
    "    \n",
    "    # add cell types\n",
    "    mask = (cf['key'] == key)\n",
    "    ctdf = cf[mask]\n",
    "    \n",
    "    ctdf = ctdf.drop_duplicates(subset=['cellType', 'clusterId'])\n",
    "    ctMapper = dict(zip(ctdf['clusterId'], ctdf['cellType']))\n",
    "    \n",
    "    clusterCount['cellType'] = clusterCount['clusterId'].map(ctMapper)\n",
    "    \n",
    "    grped = clusterCount.groupby('cellType')['proportion'].sum()\n",
    "    grped = grped.sort_index(ascending=False).reset_index()\n",
    "    \n",
    "    grped['color'] = grped['cellType'].map(colordict)\n",
    "    grped['key'] = key\n",
    "    \n",
    "    grped = grped.sort_values(by='cellType')\n",
    "    print(grped['color'].to_list())\n",
    "    print()\n",
    "    \n",
    "    res.append(grped)\n",
    "    \n",
    "    axs[i].barh(grped['cellType'],\n",
    "             grped['proportion'],\n",
    "             color=grped['color'],\n",
    "             edgecolor='k')\n",
    "    \n",
    "    axs[i].set_xlabel('Proportion of Cells')\n",
    "    axs[i].set_title(key)\n",
    "    axs[i].set_xlim([0, 1])\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "\n",
    "res = pd.concat(res,ignore_index=True)\n",
    "res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f13ba0e2-b958-4d03-9f7b-250c618696e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae78389f-fe7e-491d-b025-6157de1635ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = res.sort_values(by='key', ascending=False)\n",
    "res2 = pd.pivot_table(res, values='proportion', index=['cellType'], columns='key').reset_index()\n",
    "res2 = res2.set_index('cellType')\n",
    "# res2.columns = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 3, 3\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "# res = res.set_index('cellType')\n",
    "res2.T.plot(kind='bar', stacked=True, edgecolor='k', color=['#9e0142', '#e95c47', '#fdbf6f', '#ffffbe', '#bfe5a0', '#54aead', '#5e4fa2'])\n",
    "plt.legend(bbox_to_anchor=(1.05, 1.025))\n",
    "plt.ylabel('Proportion of Cells')\n",
    "_ = plt.xticks(rotation=0)\n",
    "plt.xlabel(\"\")\n",
    "# plt.title(\"scRNA-seq\")\n",
    "plt.gca().invert_xaxis()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6509f195-8ca8-47f7-b40c-450420427d43",
   "metadata": {},
   "outputs": [],
   "source": [
    "res2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f23ca99-e03b-4ba3-abfa-70cbdac3e85b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e977d9d9-6188-4657-b84d-b70cac9eb445",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0969ba7c-1ac1-4280-ae93-5a3622ef4390",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e9770c-0dab-4f92-87c5-59162e88e6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" a pooled approach \"\"\"\n",
    "\n",
    "allGenes = []\n",
    "clusterIdList = []\n",
    "colors = []\n",
    "cellTypes = []\n",
    "cellIds = []\n",
    "\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    data = rna[key]\n",
    "    df = data.to_df()\n",
    "    allGenes += df.columns.to_list()\n",
    "    \n",
    "    allKeys = list(data.obs.keys())\n",
    "    allKeys.remove('n_genes')\n",
    "    runKey = allKeys[0]\n",
    "    \n",
    "    clusterIds = data.obs[runKey].astype(int) + 1  # note that these are zero-indexed!\n",
    "    clusterIdList += list(clusterIds)\n",
    "    \n",
    "    cellIds += list(df.index)\n",
    "    \n",
    "    # get coloring\n",
    "    c = ['k'] * len(df)\n",
    "    cellType = ['NULL'] * len(df)\n",
    "    handles = []\n",
    "    \n",
    "    for ctype in ctypes:\n",
    "        mask = (cf['cellType'] == ctype) & (cf['key'] == key)\n",
    "        cids = cf[mask]['clusterId'].unique()\n",
    "        c = np.where(clusterIds.isin(cids), colordict[ctype], c)\n",
    "        cellType = np.where(clusterIds.isin(cids), ctype, cellType)\n",
    "        \n",
    "    colors += list(c)\n",
    "    cellTypes += list(cellType)\n",
    "    \n",
    "\n",
    "    # only consider genes present at all time points\n",
    "allGenes = [k for (k,v) in Counter(allGenes).items() if v > 2]\n",
    "\n",
    "keyVector = []\n",
    "dfList = []\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    data = rna[key]\n",
    "    df = data.to_df()\n",
    "    df = df[allGenes]\n",
    "    \n",
    "    keyVector += ([key] * df.shape[0])\n",
    "    dfList.append(df)\n",
    "\n",
    "X = pd.concat(dfList, ignore_index=True)\n",
    "X = X.to_numpy()\n",
    "\n",
    "\"\"\" PCA \"\"\"\n",
    "r = 10\n",
    "pca = PCA(n_components=r)\n",
    "pcaModel = pca.fit(X)\n",
    "Xpca = pca.fit_transform(X)\n",
    "\n",
    "\"\"\" UMAP \"\"\"\n",
    "\n",
    "# plotting params\n",
    "n_components = 2 # should always be 2\n",
    "n_neighbors = 50\n",
    "min_dist = 0.25\n",
    "metric= \"euclidean\"\n",
    "\n",
    "umap_reducer = umap.UMAP(n_components=n_components,\n",
    "                         n_neighbors=n_neighbors,\n",
    "                         min_dist=min_dist,\n",
    "                         metric=metric)\n",
    "\n",
    "ef = umap_reducer.fit_transform(Xpca) # we need embedding by cell\n",
    "ef = pd.DataFrame(ef)\n",
    "ef.columns = ['x', 'y']\n",
    "ef['key'] = keyVector\n",
    "ef['clusterId'] = clusterIdList\n",
    "ef['colors'] = colors\n",
    "ef['cellTypes'] = cellTypes\n",
    "ef['cellId'] = cellIds\n",
    "ef['cellId'] = ef['cellId'].apply(lambda x: x.replace(\"-\", \".\"))\n",
    "\n",
    "print(ef['key'].value_counts(normalize=False))\n",
    "print()\n",
    "print(ef['cellTypes'].value_counts(normalize=True))\n",
    "print()\n",
    "ef.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d9efe83-736d-4576-bbcf-56b8c28b35e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# outpath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/combinedEmbedding.csv\"\n",
    "\n",
    "# ef.to_csv(outpath, index=False)\n",
    "# print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "499c9850-22a3-4997-83dc-d7d35815ccc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88d96b71-5b9b-44cf-a579-8ab6daa622a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 6, 6\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "pdf = ef.sort_values(by='cellTypes')\n",
    "\n",
    "plt.scatter(pdf['x'], \n",
    "            pdf['y'], \n",
    "            ec='k', \n",
    "            lw=0.25,\n",
    "            c=pdf['colors'],\n",
    "            s=6)\n",
    "\n",
    "\n",
    "plt.legend(handles=handles, bbox_to_anchor=(1.35, 1.015))\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.xlabel(\"UMAP 1\")\n",
    "plt.ylabel(\"UMAP 2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b5e8d29-3825-4391-8aed-c1c64e8efab1",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 6, 6\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "pdf = ef.sort_values(by='cellTypes')\n",
    "\n",
    "mask1 = (pdf['cellTypes'] == 'Dendritic cells')\n",
    "\n",
    "plt.scatter(pdf['x'][mask1], \n",
    "            pdf['y'][mask1], \n",
    "            lw=0.25,\n",
    "            ec='purple',\n",
    "            c='purple',\n",
    "            alpha=0.9,\n",
    "            s=2)\n",
    "\n",
    "\n",
    "mask2 = (pdf['cellTypes'] == 'Macrophages')\n",
    "\n",
    "plt.scatter(pdf['x'][mask2], \n",
    "            pdf['y'][mask2], \n",
    "            lw=0.25,\n",
    "            ec='orange',\n",
    "            c='orange',\n",
    "            alpha=0.9,\n",
    "            s=2)\n",
    "\n",
    "\n",
    "mask3 = (pdf['cellTypes'] == 'B cells')\n",
    "\n",
    "plt.scatter(pdf['x'][mask3], \n",
    "            pdf['y'][mask3], \n",
    "            lw=0.25,\n",
    "            ec='green',\n",
    "            c='green',\n",
    "            alpha=0.9,\n",
    "            s=2)\n",
    "\n",
    "mask4 = mask1 | mask2 | mask3\n",
    "plt.scatter(pdf['x'][~mask4], \n",
    "            pdf['y'][~mask4], \n",
    "            ec='k', \n",
    "            lw=0.25,\n",
    "            c='grey',\n",
    "            alpha=0.5,\n",
    "            s=2)\n",
    "\n",
    "plt.axis(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "620eb058-6c9f-44b7-aaa4-03322e4d771a",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "502f299b-d3fd-4435-96c0-a53d176d72de",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 12, 4\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "fig, axs = plt.subplots(1, 3, sharey=True)\n",
    "axs = axs.ravel()\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "\n",
    "    mask = ef['key'] == key\n",
    "    pdf = ef.sort_values(by='cellTypes')\n",
    "    \n",
    "    axs[i].scatter(pdf['x'][mask], \n",
    "                   pdf['y'][mask], \n",
    "                   ec='k', \n",
    "                   lw=0.25,\n",
    "                   c=pdf['colors'][mask],\n",
    "                   zorder=2,\n",
    "                   s=10)\n",
    "        \n",
    "    axs[i].scatter(pdf['x'][~mask], \n",
    "                   pdf['y'][~mask], \n",
    "                   ec='None', \n",
    "                   lw=0.25,\n",
    "                   c='lightgrey',\n",
    "                   zorder=1,\n",
    "                   s=10)\n",
    "\n",
    "    # axs[i].set_xlabel(\"UMAP 1\")\n",
    "    # if i == 0:\n",
    "    #     axs[i].set_ylabel(\"UMAP 2\")\n",
    "    # else:\n",
    "    #     axs[i].set_ylabel(\"\")\n",
    "    # axs[i].set_title(key)\n",
    "    axs[i].set_xticks([], [])\n",
    "    axs[i].set_yticks([], [])\n",
    "    axs[i].axis(False)\n",
    "    \n",
    "\n",
    "# plt.legend(handles=handles, bbox_to_anchor=(1.04, 1.02))\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d08e4d46-b4a4-4ec2-a1f4-e5e53e392c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "choose = 50 # number of genes to consider\n",
    "\n",
    "# load gene lists \n",
    "pandPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/panglaodb/pandb.tsv.gz\"        \n",
    "pandf = ut.read_panglaodb(pandPath)\n",
    "\n",
    "controlList = [\n",
    "    'Adipocytes',\n",
    "    'B cells', \n",
    "    'Dendritic cells',\n",
    "    'Macrophages',\n",
    "    'Monocytes', \n",
    "    'NK cells',\n",
    "    'T cells', \n",
    "]\n",
    "\n",
    "allMarkers = pandf[pandf['cell type'].isin(controlList)]\n",
    "print(pd.DataFrame(allMarkers['cell type'].value_counts()))\n",
    "print()\n",
    "\"\"\"\n",
    "Ubiquitousness Index (0-1). \n",
    "0 indicates the gene is not expressed in any cell cluster and \n",
    "1 (maximum) indicates that the gene is expressed in all cell clusters. \n",
    "\"\"\"\n",
    "\n",
    "mfList = []\n",
    "\n",
    "for ctype in controlList:\n",
    "    cf = allMarkers[allMarkers['cell type'] == ctype]\n",
    "    \n",
    "    cf = cf.sort_values(by='ubiquitousness index',\n",
    "                        ascending=True)\n",
    "    \n",
    "    mfList.append(cf[['cell type', 'gene']].head(choose))\n",
    "\n",
    "\n",
    "markerGenes = pd.concat(mfList, ignore_index=True)\n",
    "\n",
    "\"\"\"NOTE: RENAME THE STROMAL CELL POPULATION\"\"\"\n",
    "markerGenes.loc[markerGenes['cell type'] == 'Adipocytes', 'cell type'] = 'Stromal cells'\n",
    "allMarkers.loc[allMarkers['cell type'] == 'Adipocytes', 'cell type'] = 'Stromal cells'\n",
    "\n",
    "print()\n",
    "print(markerGenes.head())\n",
    "\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f826bcd-1642-46ad-8829-37cc16e5e0e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf = ef.sort_values(by='cellTypes')\n",
    "pdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b5e8876-f528-4a70-a16a-651653915ea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = []\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    # get RNA Seq data\n",
    "    rf = rna[key].to_df()\n",
    "    rf.index = rf.index.str.replace(\"-\", \".\")\n",
    "    \n",
    "    for c in sorted(ef['cellTypes'].unique()):\n",
    "        # get the cell type assignments\n",
    "        mask = (ef['key'] == key) & (ef['cellTypes'] == c)\n",
    "        pdf = ef[mask].reset_index(drop=True)\n",
    "        \n",
    "        # get the marker genes\n",
    "        mGenes = allMarkers[allMarkers['cell type'] == c]['gene'].to_list()\n",
    "        mGenes = [x for x in mGenes if x in rf.columns]\n",
    "        \n",
    "        # make cell-type specific boolean matrix \n",
    "        cf = rf[mGenes].astype(bool).astype(int)\n",
    "        \n",
    "        # target matrix for cell type predictions \n",
    "        y_true = rf.index.isin(pdf['cellId']).astype(int)\n",
    "        \n",
    "        for gene in mGenes:\n",
    "            y_pred = cf[gene]\n",
    "            \n",
    "            tn, fp, fn, tp = metrics.confusion_matrix(y_true, y_pred).ravel()\n",
    "            \n",
    "            row = {\n",
    "                'key' : key,\n",
    "                'cellType' : c,\n",
    "                'color' : colordict[c],\n",
    "                'gene' : gene,\n",
    "                'nPred' : np.sum(y_true),\n",
    "                'true_positive' : tp,\n",
    "                'true_negative' : tn,\n",
    "                'false_positive' : fp,\n",
    "                'false_negative' : fn,\n",
    "                'sensitivity' : tp / (tp + fn),\n",
    "                'specificity' : tn / (tn + fp),\n",
    "                'precision' : tp / (tp + fp),\n",
    "                'F1' : (2*tp) / ((2*tp) + fp + fn),\n",
    "            }\n",
    "            \n",
    "            res.append(row)\n",
    "            \n",
    "            # break\n",
    "        # break\n",
    "    # break\n",
    "    \n",
    "res = pd.DataFrame(res)\n",
    "res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c093d538-2797-4ccc-adf8-08459b01b0d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 8, 3.5\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "fig, axs = plt.subplots(1, 3, sharey=True)\n",
    "axs = axs.ravel()\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    \n",
    "    mask = (res['key'] == key)\n",
    "\n",
    "    sns.boxplot(data=res[mask], \n",
    "                    y='cellType',\n",
    "                    x='precision',\n",
    "                    showfliers=False, \n",
    "                    palette=colordict.values(),\n",
    "                    # s=5,\n",
    "                    ax=axs[i])\n",
    "\n",
    "    axs[i].set_ylabel(\"\")\n",
    "    axs[i].set_title(key)\n",
    "    if i == 1:\n",
    "        axs[i].set_xlabel(\"Precision\")\n",
    "    else:\n",
    "        axs[i].set_xlabel(\"\")\n",
    "    axs[i].legend().remove()\n",
    "    axs[i].set_xlim([-0.1, 1.1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f08846f9-f8b0-43de-a982-018a01cfd0a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = []\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    # get RNA Seq data\n",
    "    rf = rna[key].to_df()\n",
    "    rf.index = rf.index.str.replace(\"-\", \".\")\n",
    "    \n",
    "    for c in sorted(ef['cellTypes'].unique()):\n",
    "        # get the cell type assignments\n",
    "        \n",
    "        kdf = ef[ef['key'] == key]\n",
    "        cellIds = kdf[(kdf['cellTypes'] == c)]['cellId'].to_list()\n",
    "        \n",
    "        # get the marker genes\n",
    "        mGenes = allMarkers[allMarkers['cell type'] == c]['gene'].to_list()\n",
    "        mGenes = [x for x in mGenes if x in rf.columns]\n",
    "        cf = rf[mGenes].mean(axis=1).reset_index(drop=False)\n",
    "        cf.columns = ['cellId', 'Mean Expression']\n",
    "        \n",
    "        mask = cf['cellId'].isin(cellIds)\n",
    "        \n",
    "        meanGroup = cf[mask]['Mean Expression']\n",
    "        meanOther = cf[~mask]['Mean Expression']\n",
    "        \n",
    "        lfc = np.log2(meanGroup.mean()) - np.log2(meanOther.mean())\n",
    "        \n",
    "        score, pval = scipy.stats.ranksums(meanGroup,\n",
    "                                           meanOther,\n",
    "                                           alternative='two-sided')\n",
    "                \n",
    "        row = {\n",
    "            'key' : key,\n",
    "            'cellType' : c,\n",
    "            'color' : colordict[c],\n",
    "            'pval' : pval,\n",
    "            'score' : score,\n",
    "            'meanGroup' : meanGroup.mean(),\n",
    "            'meanOther' : meanOther.mean(),\n",
    "            'log2foldchange' : lfc\n",
    "        }\n",
    "        res.append(row)\n",
    "\n",
    "    \n",
    "res = pd.DataFrame(res)\n",
    "res.head()\n",
    "\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 9, 3\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "sns.barplot(data=res, \n",
    "            x='cellType',\n",
    "            y='log2foldchange',\n",
    "            hue='key',\n",
    "            edgecolor='k', \n",
    "            palette=['lightgrey','darkgrey',  'grey' ])\n",
    "\n",
    "plt.xlabel(\"\")\n",
    "plt.ylabel(\"Fold Change (log2)\")\n",
    "plt.legend(loc='upper left')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "155adb38-f58d-47a0-b8b0-be48073516ab",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "896a177d-1d5e-45de-a8ba-1796a8fca1fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b273a51a-e55f-456b-86be-a6d8893a3ae1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2446e2ac-0b4e-4013-b046-f519038ce85b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cTypes = [\n",
    "#     'B cells',\n",
    "#     'Dendritic cells',\n",
    "#     'Macrophages',\n",
    "#     'Monocytes',\n",
    "#     'NK cells',\n",
    "#     # 'Stromal cells',\n",
    "#     'T cells',\n",
    "# ]\n",
    "\n",
    "# cmap = 'viridis'\n",
    "\n",
    "# res = []\n",
    "\n",
    "# for i, key in enumerate(keys):\n",
    "\n",
    "#     mask = ef['key'] == key\n",
    "#     pdf = ef.sort_values(by='cellTypes')\n",
    "    \n",
    "#     # get RNA Seq data\n",
    "#     rf = rna[key].to_df()\n",
    "#     rf.index = rf.index.str.replace(\"-\", \".\")\n",
    "#     res.append(rf)\n",
    "    \n",
    "# # build the datadframe\n",
    "# res = pd.concat(res)\n",
    "# res = res.fillna(0)\n",
    "\n",
    "\n",
    "# for c in cTypes:\n",
    "    \n",
    "#     plt.rcParams['figure.dpi'] = 300\n",
    "#     plt.rcParams['figure.figsize'] = 8, 4\n",
    "#     plt.rcParams['figure.facecolor'] = 'w'\n",
    "    \n",
    "#     mGenes = allMarkers[allMarkers['cell type'] == c]['gene'].to_list()\n",
    "#     mGenes = [x for x in mGenes if x in res.columns]\n",
    "    \n",
    "#     trf = pd.DataFrame(res[mGenes].mean(axis=1)).reset_index(drop=False)\n",
    "#     trf.columns = ['cellId', 'exp']\n",
    "\n",
    "#     ppdf = pd.merge(pdf, trf, \n",
    "#                     left_on='cellId',\n",
    "#                     right_on='cellId',\n",
    "#                     how='left')\n",
    "    \n",
    "#     fig, axs = plt.subplots(1, 2, sharey=True)\n",
    "#     axs = axs.ravel()\n",
    "    \n",
    "#     mask = (ppdf['cellTypes'] == c)\n",
    "    \n",
    "#     axs[0].scatter(ppdf['x'][mask], \n",
    "#                    ppdf['y'][mask], \n",
    "#                    ec='k', \n",
    "#                    lw=0.25,\n",
    "#                    c=ppdf['colors'][mask],\n",
    "#                    zorder=2,\n",
    "#                    s=6)\n",
    "    \n",
    "#     axs[0].scatter(ppdf['x'][~mask], \n",
    "#                    ppdf['y'][~mask], \n",
    "#                    ec='lightgrey', \n",
    "#                    lw=0.25,\n",
    "#                    c='lightgrey',\n",
    "#                    zorder=1,\n",
    "#                    s=6)\n",
    "    \n",
    "#     axs[0].set_title(f'{c}')\n",
    "#     axs[0].set_xticks([])\n",
    "#     axs[0].set_yticks([])\n",
    "#     axs[0].set_xlabel(\"UMAP 1\")\n",
    "#     axs[0].set_ylabel(\"UMAP 2\")\n",
    "    \n",
    "#     axs[1].scatter(ppdf['x'], \n",
    "#                    ppdf['y'], \n",
    "#                    ec='k', \n",
    "#                    lw=0.25,\n",
    "#                    c=ppdf['exp'],\n",
    "#                    cmap=cmap,\n",
    "#                    zorder=2,\n",
    "#                    s=6)\n",
    "    \n",
    "#     axs[1].set_xticks([])\n",
    "#     axs[1].set_yticks([])\n",
    "#     axs[1].set_xlabel(\"UMAP 1\")\n",
    "#     axs[1].set_title(f'Marker Gene Expression')\n",
    "    \n",
    "# #     norm = plt.Normalize(ppdf['exp'].min(),ppdf['exp'].max())\n",
    "# #     sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "# #     sm.set_array([])\n",
    "# #     cbar = plt.colorbar(sm, \n",
    "# #                         ax=axs[1], \n",
    "# #                         fraction=0.046, \n",
    "# #                         pad=0.04, \n",
    "# #                         label=f'Mean Expression')\n",
    "    \n",
    "# #     cbar.ax.set_yticklabels([])\n",
    "# #     cbar.ax.set_title('High', fontsize=10)\n",
    "# #     cbar.ax.set_xlabel('Low', fontsize=10)\n",
    "        \n",
    "    \n",
    "#     plt.tight_layout()\n",
    "#     plt.show()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01d9ce2d-7565-4ea6-85b8-f9cb0728acf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def makeColorbar(cmap, width, hieght, title, orientation, tickLabels):\n",
    "#     a = np.array([[0,1]])\n",
    "#     plt.figure(figsize=(width, hieght))\n",
    "#     img = plt.imshow(a, cmap=cmap)\n",
    "#     plt.gca().set_visible(False)\n",
    "#     cax = plt.axes([0.1, 0.2, 0.8, 0.6])\n",
    "#     ticks = np.linspace(0,1 , len(tickLabels))\n",
    "#     cbar = plt.colorbar(orientation=orientation, \n",
    "#                         cax=cax, \n",
    "#                         label=title,\n",
    "#                         ticks=ticks)\n",
    "\n",
    "#     if orientation == 'vertical':\n",
    "#         cbar.ax.set_yticklabels(tickLabels)\n",
    "#     else:\n",
    "#         cbar.ax.set_xticklabels(tickLabels)\n",
    "    \n",
    "# makeColorbar(\"viridis\", 0.3, 2, 'Mean Expression', 'vertical', [\"Low\", \"High\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96a8a15e-38fb-410b-a9e1-669028fb8d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e1d2a93-b8dc-4fdf-bd43-344d6915e8b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 5, 5\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "print(ef['cellTypes'].unique())\n",
    "\n",
    "query = 'Macrophages'\n",
    "n = 6\n",
    "cmap = \"RdYlBu\"\n",
    "colors = list(reversed(ut.ncolor(n, cmap=cmap)[:3]))\n",
    "# colors = ['#2329e8', \"#f2f252\", \"#e82323\"]\n",
    "keyMap = dict(zip(ef['key'].unique(), colors))\n",
    "\n",
    "keyHandles = []\n",
    "\n",
    "for k, c in keyMap.items():        \n",
    "    handle = mpatches.Patch(color=c, \n",
    "                            ec='k', \n",
    "                            label=k)\n",
    "    keyHandles.append(handle)\n",
    "\n",
    "pdf = ef.copy()\n",
    "mask = pdf['cellTypes'] == query\n",
    "\n",
    "zorder = {\n",
    "    'ND' : 4,\n",
    "    'HFD8' : 2,\n",
    "    'HFD14' : 3\n",
    "}\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    \n",
    "    mask = (pdf['cellTypes'] == query) & (pdf['key'] == key)\n",
    "    \n",
    "    kpdf = pdf[mask]\n",
    "    \n",
    "\n",
    "    plt.scatter(kpdf['x'], \n",
    "                kpdf['y'], \n",
    "                ec='k', \n",
    "                lw=0.25,\n",
    "                c=colors[i],\n",
    "                # alpha=0.7,\n",
    "                zorder=zorder[key],\n",
    "                # zorder=2,\n",
    "                s=10)\n",
    "    \n",
    "\n",
    "plt.scatter(pdf['x'][~mask], \n",
    "            pdf['y'][~mask], \n",
    "            ec='None', \n",
    "            lw=0.25,\n",
    "            c='lightgrey',\n",
    "            zorder=1,\n",
    "            s=10)\n",
    "\n",
    "\n",
    "plt.title(query)\n",
    "plt.ylim([-3, 7.3])\n",
    "plt.xlim([-5, 7.5])\n",
    "plt.legend(handles=keyHandles, bbox_to_anchor=(1.0, 0.81))\n",
    "plt.xticks([], [])\n",
    "plt.yticks([], [])\n",
    "plt.xlabel(\"UMAP 1\")\n",
    "plt.ylabel(\"UMAP 2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4396d230-6efd-43b5-941f-89dc9a5f537e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ut.ncolor(3, cmap=\"Spectral_r\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6342781-f9f6-4bd2-a9df-a98897eca8b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#     # print(df.head())\n",
    "    \n",
    "#     var = df.std(axis=0)\n",
    "#     mask = var > minStd\n",
    "#     print(f\"{mask.sum()} ({mask.sum() / len(mask):.3f}) genes with std > {minStd}\")\n",
    "    \n",
    "#     df = df[df.columns[np.where(mask)]]\n",
    "    \n",
    "    \n",
    "#     # df = df[[mask.index.to_lis/t()]]\n",
    "    \n",
    "#     # print(df[['MRPL15']].head())\n",
    "# #     \n",
    "#     # sc.pp.highly_variable_genes(data, \n",
    "#     #                             min_mean=0.0125, \n",
    "#     #                             max_mean=3, \n",
    "#     #                             min_disp=0.5)\n",
    "    \n",
    "#     break\n",
    "        \n",
    "#     # get the run key\n",
    "#     allKeys = list(data.obs.keys())\n",
    "#     allKeys.remove('n_genes')\n",
    "#     runKey = allKeys[0]\n",
    "    \n",
    "#     clusterIds = data.obs[runKey].astype(int) + 1  # note that these are zero-indexed!\n",
    "#     embedding = data.obsm['X_umap']\n",
    "\n",
    "#     # get coloring\n",
    "#     c = ['k'] * len(embedding)\n",
    "#     cellType = ['NULL'] * len(embedding)\n",
    "#     handles = []\n",
    "    \n",
    "#     for ctype in ctypes:\n",
    "#         mask = (cf['cellType'] == ctype) & (cf['key'] == key)\n",
    "#         cids = cf[mask]['clusterId'].unique()\n",
    "#         c = np.where(clusterIds.isin(cids), colordict[ctype], c)\n",
    "#         cellType = np.where(clusterIds.isin(cids), ctype, cellType)\n",
    "        \n",
    "#     pdf = pd.DataFrame({'x' : embedding[:, 0], \n",
    "#                         'y' : embedding[:, 1]})\n",
    "    \n",
    "#     pdf['cellType'] = cellType\n",
    "#     pdf['Color'] = c\n",
    "#     pdf['clusterId'] = list(clusterIds)\n",
    "    \n",
    "#     pdf['key'] = key\n",
    "    \n",
    "#     dfList.append(pdf)\n",
    "    \n",
    "    \n",
    "    \n",
    "# pdf = pd.concat(dfList, ignore_index=True)\n",
    "\n",
    "\n",
    "# plt.rcParams['figure.dpi'] = 300\n",
    "# plt.rcParams['figure.figsize'] = 15, 15\n",
    "# plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "\n",
    "# sns.scatterplot(data=pdf, \n",
    "#                 x='x', \n",
    "#                 y='y', \n",
    "#                 ec='k', \n",
    "#                 hue='cellType', \n",
    "#                 palette='Spectral')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea9c66fc-eff1-4e3f-a779-11afa3daae45",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
