{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "06f3c223-d25d-49da-92ea-ab3c1964b282",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-12-19 16:30:43.598425: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2022-12-19 16:30:43.713099: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/slurm/lib64:\n",
      "2022-12-19 16:30:43.713118: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\n",
      "2022-12-19 16:30:43.739472: E tensorflow/stream_executor/cuda/cuda_blas.cc:2981] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered\n",
      "2022-12-19 16:30:45.372596: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/slurm/lib64:\n",
      "2022-12-19 16:30:45.372686: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer_plugin.so.7'; dlerror: libnvinfer_plugin.so.7: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/slurm/lib64:\n",
      "2022-12-19 16:30:45.372692: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Cannot dlopen some TensorRT libraries. If you would like to use Nvidia GPU with TensorRT, please make sure the missing libraries mentioned above are installed properly.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib \n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import squidpy as sq\n",
    "from sklearn import metrics\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "import tabulate\n",
    "import itertools\n",
    "from mpl_toolkits import mplot3d\n",
    "import matplotlib.patches as mpatches\n",
    "import umap\n",
    "from importlib import reload\n",
    "from scipy import stats\n",
    "import matplotlib.image as mpimg\n",
    "import gget\n",
    "\n",
    "# locals\n",
    "import utils as ut\n",
    "reload(ut)\n",
    "\n",
    "sc.settings.verbosity = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "80b471a9-4699-4c6e-8618-7fe3732d2ee2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(427882, 7)\n",
      "['Macrophages' 'B cells' 'Monocytes' 'NK cells' 'Dendritic cells'\n",
      " 'T cells' 'Stromal cells']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>scores</th>\n",
       "      <th>pvals</th>\n",
       "      <th>logfoldchanges</th>\n",
       "      <th>clusterId</th>\n",
       "      <th>key</th>\n",
       "      <th>cellType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>H2-EB1</td>\n",
       "      <td>36.343353</td>\n",
       "      <td>4.575346e-181</td>\n",
       "      <td>3.564091</td>\n",
       "      <td>1</td>\n",
       "      <td>ND</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>H2-AA</td>\n",
       "      <td>35.311523</td>\n",
       "      <td>4.761517e-183</td>\n",
       "      <td>3.430534</td>\n",
       "      <td>1</td>\n",
       "      <td>ND</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>H2-AB1</td>\n",
       "      <td>34.411163</td>\n",
       "      <td>2.815112e-177</td>\n",
       "      <td>3.234597</td>\n",
       "      <td>1</td>\n",
       "      <td>ND</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CD74</td>\n",
       "      <td>33.910587</td>\n",
       "      <td>2.451588e-177</td>\n",
       "      <td>3.302062</td>\n",
       "      <td>1</td>\n",
       "      <td>ND</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>H2-DMA</td>\n",
       "      <td>31.695400</td>\n",
       "      <td>2.991793e-123</td>\n",
       "      <td>2.667091</td>\n",
       "      <td>1</td>\n",
       "      <td>ND</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene     scores          pvals  logfoldchanges  clusterId key  \\\n",
       "0  H2-EB1  36.343353  4.575346e-181        3.564091          1  ND   \n",
       "1   H2-AA  35.311523  4.761517e-183        3.430534          1  ND   \n",
       "2  H2-AB1  34.411163  2.815112e-177        3.234597          1  ND   \n",
       "3    CD74  33.910587  2.451588e-177        3.302062          1  ND   \n",
       "4  H2-DMA  31.695400  2.991793e-123        2.667091          1  ND   \n",
       "\n",
       "      cellType  \n",
       "0  Macrophages  \n",
       "1  Macrophages  \n",
       "2  Macrophages  \n",
       "3  Macrophages  \n",
       "4  Macrophages  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusterPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/scanpy/clusters.csv\"\n",
    "cf = pd.read_parquet(clusterPath)\n",
    "print(cf.shape)\n",
    "print(cf['cellType'].unique())\n",
    "cf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "332a9137-6b25-462f-8c7b-45b07a21f575",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading /nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/SPT/HFD14/outs/filtered_feature_bc_matrix.h5\n",
      " (0:00:00)\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "... as `zero_center=True`, sparse input is densified and may lead to large memory consumption\n",
      "reading /nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/SPT/ND/outs/filtered_feature_bc_matrix.h5\n",
      " (0:00:00)\n",
      "filtered out 1 cells that have less than 10 counts\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "... as `zero_center=True`, sparse input is densified and may lead to large memory consumption\n",
      "reading /nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/SPT/HFD8/outs/filtered_feature_bc_matrix.h5\n",
      " (0:00:00)\n",
      "filtered out 2 cells that have less than 10 counts\n",
      "normalizing counts per cell\n",
      "    finished (0:00:00)\n",
      "... as `zero_center=True`, sparse input is densified and may lead to large memory consumption\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "dict_keys(['HFD14', 'ND', 'HFD8'])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dirPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/SPT/\"\n",
    "\n",
    "min_counts = 10\n",
    "\n",
    "adata = {}\n",
    "\n",
    "for f in os.listdir(dirPath):\n",
    "    \n",
    "    if not f == \"all\":\n",
    "        key = f\n",
    "        mtxPath = f\"{dirPath}{f}/outs/\"\n",
    "        data = sc.read_visium(mtxPath)\n",
    "        data.var_names_make_unique()\n",
    "        data.var_names = [x.upper() for x in data.var_names]\n",
    "        \n",
    "        \"\"\" filter the SPT data\"\"\"\n",
    "        sc.pp.filter_cells(data, min_counts=min_counts)\n",
    "        sc.pp.normalize_total(data, target_sum=1e4) # Normalize each cell by total counts over all genes\n",
    "        sc.pp.log1p(data) # Logarithmize data via `X = \\log(X + 1)`,\n",
    "        sc.pp.scale(data) # unit variance and zero mean\n",
    "        \n",
    "        adata[key] = data\n",
    "    \n",
    "adata.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7c1985bb-d8d5-4182-b879-557b7b354ce3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "dirPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/scanpy/\"\n",
    "\n",
    "rna = {}\n",
    "\n",
    "for f in os.listdir(dirPath):\n",
    "    \n",
    "    if f.endswith('.h5ad'):\n",
    "        key = f.split(\".\")[0]\n",
    "        fullPath = f\"{dirPath}{f}\"\n",
    "        data = sc.read(fullPath)\n",
    "        rna[key] = data\n",
    "    \n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b3d62400-24f9-488c-81b0-50323b33d38e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "# load cluster assignments\n",
    "cdf = pd.read_csv(\"clusterAssignments.csv\")\n",
    "cid = {}\n",
    "\n",
    "for key in cdf['key'].unique():\n",
    "    t = cdf[cdf['key'] == key]\n",
    "    cid[key] = t\n",
    "    \n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "efdd4422-c142-4919-b8a9-9b844a0f03ba",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'outpdir' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_3830856/790694280.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     41\u001b[0m     \u001b[0mclusterIds\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_parquet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutpath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 43\u001b[0;31m     \u001b[0moutpath\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34mf\"{outpdir}{key}_coords.pq\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     44\u001b[0m     \u001b[0mcoords\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_parquet\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutpath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'outpdir' is not defined"
     ]
    }
   ],
   "source": [
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "outdir = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/cleanData/\"\n",
    "\n",
    "clusters = []\n",
    "rnas = []\n",
    "spts = []\n",
    "\n",
    "for key in keys:\n",
    "    spt = adata[key]\n",
    "    scr = rna[key]\n",
    "    \n",
    "    # get coordinates\n",
    "    coords = pd.DataFrame(spt.obsm['spatial'])\n",
    "    coords.columns = ['x', 'y']\n",
    "    \n",
    "    \n",
    "    # get the run key\n",
    "    allKeys = list(scr.obs.keys() )\n",
    "    allKeys.remove('n_genes')\n",
    "    runKey = allKeys[0]\n",
    "    clusterIds = scr.obs[runKey].astype(int) + 1  # note that these are zero-indexed!\n",
    "    clusterIds = pd.DataFrame(clusterIds)\n",
    "    clusterIds = clusterIds.reset_index(drop=False)\n",
    "    clusterIds.columns = ['cellId', 'clusterId']\n",
    "    \n",
    "    clusterAssignments = cid[key][['clusterId', 'cellType']]\n",
    "    \n",
    "    clusterIds = pd.merge(clusterIds, \n",
    "                          clusterAssignments,\n",
    "                          how='left',\n",
    "                          left_on='clusterId',\n",
    "                          right_on='clusterId')\n",
    "    \n",
    "    src = scr.to_df()\n",
    "    spt = spt.to_df()\n",
    "    \n",
    "    coords.index = spt.index\n",
    "    \n",
    "    outpath = f\"{outdir}{key}_clusters.pq\"\n",
    "    clusterIds.to_parquet(outpath, index=False)\n",
    "    \n",
    "    outpath = f\"{outdir}{key}_coords.pq\"\n",
    "    coords.to_parquet(outpath, index=True)\n",
    "    \n",
    "    outpath = f\"{outdir}{key}_rna.pq\"\n",
    "    src.to_parquet(outpath, index=True)\n",
    "    \n",
    "    outpath = f\"{outdir}{key}_spt.pq\"\n",
    "    spt.to_parquet(outpath, index=True)\n",
    "    \n",
    "    print('done', key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a55ba4d5-2b75-4358-9a04-8e51ebb663ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44315c15-ac5e-4974-9184-3f5248a6dcc8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74583ab1-55fc-4479-b8f3-c1eba6daffb6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69cf6adc-10e4-4257-b086-7a65501b1d94",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75fbfad7-31ae-4c0b-989e-9e416290e59a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getImage(data):\n",
    "    \"\"\"A function to get image from the spatial data \"\"\"\n",
    "    from scanpy.pl._tools import scatterplots as spt    \n",
    "    sptData = data.uns['spatial']\n",
    "    \n",
    "    libraryId = list(sptData.keys())[0]\n",
    "    sptData = sptData[libraryId]\n",
    "    \n",
    "    # get params\n",
    "    img = sptData['images']['hires']\n",
    "    scale_factor = sptData['scalefactors']['tissue_hires_scalef']\n",
    "    return img, scale_factor\n",
    "\n",
    "\n",
    "trim = {\n",
    "    'ND' : {'xlim' :  [250, 1500], 'ylim' : [1600, 650]},\n",
    "    'HFD8' : {'xlim' :  [190, 1350], 'ylim' : [1680, 400]},\n",
    "    'HFD14' : {'xlim' :  [250, 1400], 'ylim' : [1650, 450]},\n",
    "}\n",
    "\n",
    "print(trim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40a81ade-c202-4e27-aac0-be91745a318b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctypes = sorted(cf['cellType'].unique())\n",
    "nTypes = len(ctypes)\n",
    "colorList = ut.ncolor(nTypes, cmap='Spectral')\n",
    "colorDict = dict(zip(ctypes, colorList))\n",
    "\n",
    "cf['color'] = cf['cellType'].map(colorDict)\n",
    "cf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b72054fd-455d-4377-81bd-32be430b2127",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 10, 10\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "lft = 1\n",
    "nGene = 30\n",
    "alpha = 0.05\n",
    "q = 20\n",
    "layer = 6 # reverse ordered from 6 to 0 with 6 == 1 first layer\n",
    "\n",
    "for key in keys:\n",
    "    data = adata[key]\n",
    "    # get the image and the scale factors\n",
    "    img, scale_factor = getImage(data)\n",
    "    df = data.to_df()\n",
    "    \n",
    "    # get the image coordinates and scale them\n",
    "    coords = data.obsm['spatial']\n",
    "    x = coords[:, 0]\n",
    "    y = coords[:, 1]\n",
    "    x = x * scale_factor \n",
    "    y = y * scale_factor \n",
    "    \n",
    "    # get marker genes for cell types\n",
    "    rf = cf[cf['key'] == key]\n",
    "    rf = rf[rf['pvals'] <= alpha].reset_index(drop=True)\n",
    "    \n",
    "    rf = rf.groupby(['gene', 'cellType']).agg(\n",
    "        clusterCount = ('clusterId', 'count'),\n",
    "        meanLFC = ('logfoldchanges', 'mean')\n",
    "    ).reset_index(drop=False)\n",
    "    \n",
    "    rf = rf[rf['meanLFC'] > lft].reset_index(drop=True)        \n",
    "    rf = rf.set_index('gene')\n",
    "    rf = rf.groupby(['cellType'])['clusterCount'].nlargest(nGene).reset_index(drop=False)\n",
    "    rf['flag'] = 1.0\n",
    "    \n",
    "    # print(rf.head())\n",
    "    # print(rf['cellType'].value_counts())\n",
    "    \n",
    "    rf = pd.pivot_table(rf, values='flag',\n",
    "                        index='gene', \n",
    "                        columns='cellType')\n",
    "    \n",
    "    rf = rf.fillna(0.0)\n",
    "\n",
    "    dfT = df.T\n",
    "    # binarize the spot data\n",
    "    spots = dfT.columns\n",
    "    \n",
    "    spotDecon = pd.merge(dfT, rf,\n",
    "                         how='left',  \n",
    "                         left_on=dfT.index,\n",
    "                         right_on=rf.index,)\n",
    "    \n",
    "    spotDecon = spotDecon.fillna(0)\n",
    "    \"\"\" Deconvolute the spots \"\"\"\n",
    "    \n",
    "    spotColor = []\n",
    "    spotTypes = []\n",
    "    for spot in spots:\n",
    "        spotThresh = np.percentile(spotDecon[spot], q)\n",
    "        spotVec = np.where(spotDecon[spot] > spotThresh, 1.0, 0.0)\n",
    "        scores = []\n",
    "        for c in ctypes:\n",
    "            \n",
    "            score = metrics.jaccard_score(spotDecon[c], spotVec)\n",
    "            scores.append(score)\n",
    "        \n",
    "        indx = np.argsort(scores)\n",
    "        i = np.nonzero(indx == layer)[0][0]\n",
    "        spotType = ctypes[i]\n",
    "        spotColor.append(colorDict[spotType])\n",
    "        spotTypes.append(spotType)\n",
    "        \n",
    "    print(pd.Series(spotTypes).value_counts())\n",
    "        \n",
    "    handles = []\n",
    "    for ctype in ctypes:\n",
    "        handle = mpatches.Patch(color=colorDict[ctype], \n",
    "                                ec='k', \n",
    "                                label=ctype)\n",
    "        handles.append(handle)\n",
    "        \n",
    "    plt.scatter(x, y,\n",
    "                c=spotColor,\n",
    "                marker=\"o\",\n",
    "                edgecolor=\"None\",\n",
    "                # lw=0.1,\n",
    "                zorder=2,\n",
    "                alpha=0.7,\n",
    "                s=60)\n",
    "    \n",
    "    # plt.imshow(img.astype(float), \n",
    "    #        interpolation='none',\n",
    "    #        cmap='binary',\n",
    "    #        alpha=0.7,\n",
    "    #        zorder=1)\n",
    "    \n",
    "    xlim = trim[key]['xlim']\n",
    "    ylim = trim[key]['ylim']\n",
    "    plt.xlim(xlim)\n",
    "    plt.ylim(ylim)\n",
    "    plt.yticks([])\n",
    "    plt.xticks([])\n",
    "    \n",
    "    plt.legend(handles=handles, bbox_to_anchor=(1.04, 1.02))\n",
    "    plt.title(key)\n",
    "    plt.show()\n",
    "    \n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af02d6c2-1c8b-479e-ad8f-f0f87156f45c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6fb22a7-f01d-4631-a509-fa370747d041",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 10, 10\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "expT = 0 # gene expression threshold\n",
    "\n",
    "\n",
    "for key in keys:\n",
    "    data = adata[key]\n",
    "    # get the image and the scale factors\n",
    "    img, scale_factor = getImage(data)\n",
    "    df = data.to_df()\n",
    "    \n",
    "    # get the image coordinates and scale them\n",
    "    coords = data.obsm['spatial']\n",
    "    x = coords[:, 0]\n",
    "    y = coords[:, 1]\n",
    "    x = x * scale_factor \n",
    "    y = y * scale_factor \n",
    "    \n",
    "    spotType = []\n",
    "    \n",
    "    for idx, row in df.head(10).iterrows():\n",
    "        mask = row > expT\n",
    "        expressedGenes = df.columns[mask]\n",
    "        \n",
    "        ef = gget.enrichr(expressedGenes, database='PanglaoDB_Augmented_2021')\n",
    "        predType = ef.head(1)['path_name'].values[0]\n",
    "        spotType.append(predType)\n",
    "        \n",
    "    \n",
    "    print(pd.Series(spotType).value_counts())\n",
    "\n",
    "   \n",
    "        \n",
    "#     handles = []\n",
    "#     for ctype in ctypes:\n",
    "#         handle = mpatches.Patch(color=colorDict[ctype], \n",
    "#                                 ec='k', \n",
    "#                                 label=ctype)\n",
    "#         handles.append(handle)\n",
    "        \n",
    "#     plt.scatter(x, y,\n",
    "#                 c=spotColor,\n",
    "#                 marker=\"o\",\n",
    "#                 edgecolor=\"None\",\n",
    "#                 # lw=0.1,\n",
    "#                 zorder=2,\n",
    "#                 alpha=0.7,\n",
    "#                 s=60)\n",
    "    \n",
    "#     # plt.imshow(img.astype(float), \n",
    "#     #        interpolation='none',\n",
    "#     #        cmap='binary',\n",
    "#     #        alpha=0.7,\n",
    "#     #        zorder=1)\n",
    "    \n",
    "#     xlim = trim[key]['xlim']\n",
    "#     ylim = trim[key]['ylim']\n",
    "#     plt.xlim(xlim)\n",
    "#     plt.ylim(ylim)\n",
    "#     plt.yticks([])\n",
    "#     plt.xticks([])\n",
    "    \n",
    "#     plt.legend(handles=handles, bbox_to_anchor=(1.04, 1.02))\n",
    "#     plt.title(key)\n",
    "#     plt.show()\n",
    "    \n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6031ba-3164-4203-9c9e-f6fc62faff65",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
