{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "98642360-6da7-48b6-b03c-9b9243a4988f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib \n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "from collections import Counter\n",
    "from textwrap import wrap\n",
    "import scipy\n",
    "from mpl_toolkits.axes_grid1.axes_divider import make_axes_locatable\n",
    "from sklearn import metrics\n",
    "import gget\n",
    "import tabulate\n",
    "from sklearn.neighbors import BallTree\n",
    "from scipy.stats import gmean\n",
    "from collections import Counter\n",
    "import sys\n",
    "\n",
    "\n",
    "# locals\n",
    "sys.path.append('../')\n",
    "import utils as ut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6128443b-93e4-49eb-9efa-435593af3c63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2000, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pair.Name</th>\n",
       "      <th>Ligand.Mouse</th>\n",
       "      <th>Receptor.Mouse</th>\n",
       "      <th>Source</th>\n",
       "      <th>ManualAnnotation</th>\n",
       "      <th>Ligand.CC</th>\n",
       "      <th>Ligand.GO</th>\n",
       "      <th>Reference (PMID / KEGG)</th>\n",
       "      <th>ligand</th>\n",
       "      <th>receptor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2m-Lrp1</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>Ramilowski</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Secreted</td>\n",
       "      <td>Other</td>\n",
       "      <td>1702392,10652313, 12194978</td>\n",
       "      <td>A2M</td>\n",
       "      <td>LRP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Adgre5-Cd55</td>\n",
       "      <td>Adgre5</td>\n",
       "      <td>Cd55</td>\n",
       "      <td>Baccin</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Membrane</td>\n",
       "      <td>Other</td>\n",
       "      <td>23447688</td>\n",
       "      <td>ADGRE5</td>\n",
       "      <td>CD55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Adipoq-Adipor1</td>\n",
       "      <td>Adipoq</td>\n",
       "      <td>Adipor1</td>\n",
       "      <td>Ramilowski</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Secreted</td>\n",
       "      <td>Other</td>\n",
       "      <td>12802337</td>\n",
       "      <td>ADIPOQ</td>\n",
       "      <td>ADIPOR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Adipoq-Adipor2</td>\n",
       "      <td>Adipoq</td>\n",
       "      <td>Adipor2</td>\n",
       "      <td>Ramilowski</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Secreted</td>\n",
       "      <td>Other</td>\n",
       "      <td>12802337, 12802330</td>\n",
       "      <td>ADIPOQ</td>\n",
       "      <td>ADIPOR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Adm-Calcrl&amp;Ramp2</td>\n",
       "      <td>Adm</td>\n",
       "      <td>Calcrl&amp;Ramp2</td>\n",
       "      <td>Baccin</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Secreted</td>\n",
       "      <td>Other</td>\n",
       "      <td>10342881</td>\n",
       "      <td>ADM</td>\n",
       "      <td>CALCRL&amp;RAMP2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Pair.Name Ligand.Mouse Receptor.Mouse      Source ManualAnnotation  \\\n",
       "0          A2m-Lrp1          A2m           Lrp1  Ramilowski          Correct   \n",
       "1       Adgre5-Cd55       Adgre5           Cd55      Baccin          Correct   \n",
       "2    Adipoq-Adipor1       Adipoq        Adipor1  Ramilowski          Correct   \n",
       "3    Adipoq-Adipor2       Adipoq        Adipor2  Ramilowski          Correct   \n",
       "4  Adm-Calcrl&Ramp2          Adm   Calcrl&Ramp2      Baccin          Correct   \n",
       "\n",
       "  Ligand.CC Ligand.GO     Reference (PMID / KEGG)  ligand      receptor  \n",
       "0  Secreted     Other  1702392,10652313, 12194978     A2M          LRP1  \n",
       "1  Membrane     Other                    23447688  ADGRE5          CD55  \n",
       "2  Secreted     Other                    12802337  ADIPOQ       ADIPOR1  \n",
       "3  Secreted     Other          12802337, 12802330  ADIPOQ       ADIPOR2  \n",
       "4  Secreted     Other                    10342881     ADM  CALCRL&RAMP2  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/LR/LewisLabUSCS/Mouse/Mouse-2020-Baccin-LR-pairs.xlsx\"\n",
    "\n",
    "lr = pd.read_excel(fPath, engine=\"openpyxl\")\n",
    "print(lr.shape)\n",
    "\n",
    "lr['ligand'] = lr['Ligand.Mouse'].str.upper()\n",
    "lr['receptor'] = lr['Receptor.Mouse'].str.upper()\n",
    "lr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f465b122-a535-4b5c-a259-f6e907ab4cd6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(14836, 13)\n",
      "HFD8     6782\n",
      "ND       4068\n",
      "HFD14    3986\n",
      "Name: key, dtype: int64\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spotId</th>\n",
       "      <th>Dendritic cells</th>\n",
       "      <th>Mac4</th>\n",
       "      <th>Stromal cells</th>\n",
       "      <th>NK cells</th>\n",
       "      <th>Mac2</th>\n",
       "      <th>Monocytes</th>\n",
       "      <th>T cells</th>\n",
       "      <th>Mac5</th>\n",
       "      <th>Mac3</th>\n",
       "      <th>Mac1</th>\n",
       "      <th>B cells</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACAAGTATCTCCCA.1</td>\n",
       "      <td>0.087213</td>\n",
       "      <td>0.113186</td>\n",
       "      <td>0.270488</td>\n",
       "      <td>0.096431</td>\n",
       "      <td>0.057255</td>\n",
       "      <td>0.083130</td>\n",
       "      <td>0.047384</td>\n",
       "      <td>0.163491</td>\n",
       "      <td>0.049659</td>\n",
       "      <td>0.006606</td>\n",
       "      <td>0.025157</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACACCAATAACTGC.1</td>\n",
       "      <td>0.159247</td>\n",
       "      <td>0.129071</td>\n",
       "      <td>0.105884</td>\n",
       "      <td>0.002606</td>\n",
       "      <td>0.084753</td>\n",
       "      <td>0.088446</td>\n",
       "      <td>0.016601</td>\n",
       "      <td>0.302068</td>\n",
       "      <td>0.046115</td>\n",
       "      <td>0.038813</td>\n",
       "      <td>0.026395</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACAGCTTTCAGAAG.1</td>\n",
       "      <td>0.042437</td>\n",
       "      <td>0.128286</td>\n",
       "      <td>0.329050</td>\n",
       "      <td>0.064440</td>\n",
       "      <td>0.050805</td>\n",
       "      <td>0.049080</td>\n",
       "      <td>0.052493</td>\n",
       "      <td>0.100578</td>\n",
       "      <td>0.065773</td>\n",
       "      <td>0.025274</td>\n",
       "      <td>0.091783</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACAGGGTCTATATT.1</td>\n",
       "      <td>0.038153</td>\n",
       "      <td>0.094966</td>\n",
       "      <td>0.294255</td>\n",
       "      <td>0.055503</td>\n",
       "      <td>0.061959</td>\n",
       "      <td>0.060107</td>\n",
       "      <td>0.048176</td>\n",
       "      <td>0.127678</td>\n",
       "      <td>0.099188</td>\n",
       "      <td>0.040547</td>\n",
       "      <td>0.079468</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACAGTGTTCCTGGG.1</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.256841</td>\n",
       "      <td>0.323585</td>\n",
       "      <td>0.003329</td>\n",
       "      <td>0.015668</td>\n",
       "      <td>0.010229</td>\n",
       "      <td>0.002518</td>\n",
       "      <td>0.222850</td>\n",
       "      <td>0.000587</td>\n",
       "      <td>0.000737</td>\n",
       "      <td>0.163636</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spotId  Dendritic cells      Mac4  Stromal cells  NK cells  \\\n",
       "0  AAACAAGTATCTCCCA.1         0.087213  0.113186       0.270488  0.096431   \n",
       "1  AAACACCAATAACTGC.1         0.159247  0.129071       0.105884  0.002606   \n",
       "2  AAACAGCTTTCAGAAG.1         0.042437  0.128286       0.329050  0.064440   \n",
       "3  AAACAGGGTCTATATT.1         0.038153  0.094966       0.294255  0.055503   \n",
       "4  AAACAGTGTTCCTGGG.1         0.000020  0.256841       0.323585  0.003329   \n",
       "\n",
       "       Mac2  Monocytes   T cells      Mac5      Mac3      Mac1   B cells   key  \n",
       "0  0.057255   0.083130  0.047384  0.163491  0.049659  0.006606  0.025157  HFD8  \n",
       "1  0.084753   0.088446  0.016601  0.302068  0.046115  0.038813  0.026395  HFD8  \n",
       "2  0.050805   0.049080  0.052493  0.100578  0.065773  0.025274  0.091783  HFD8  \n",
       "3  0.061959   0.060107  0.048176  0.127678  0.099188  0.040547  0.079468  HFD8  \n",
       "4  0.015668   0.010229  0.002518  0.222850  0.000587  0.000737  0.163636  HFD8  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cardOutDir = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/CARDOuputs/\"\n",
    "\n",
    "dfList = []\n",
    "\n",
    "for f in os.listdir(cardOutDir):\n",
    "    if 'macrophage' in f:\n",
    "        fullPath = f\"{cardOutDir}{f}\"\n",
    "        key = f.split(\"_\")[0]\n",
    "        df = pd.read_csv(fullPath)\n",
    "        df = df.rename(columns={'Unnamed: 0' : 'spotId'})\n",
    "        df['key'] = key\n",
    "        dfList.append(df)\n",
    "\n",
    "df = pd.concat(dfList, ignore_index=True)\n",
    "print(f\"{df.shape=}\")\n",
    "print(df['key'].value_counts())\n",
    "print()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e331e9f0-18df-4971-8811-54fc98619d66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HFD8_macrophage_spt.csv\n",
      "HFD14_macrophage_spt.csv\n",
      "ND_macrophage_spt.csv\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "sptDir =  \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/CARDInputs/\"\n",
    "\n",
    "spt = {}\n",
    "\n",
    "for f in os.listdir(sptDir):\n",
    "    if \"spt\" in f and \"macrophage\" in f:\n",
    "        print(f)\n",
    "        fullPath = f\"{sptDir}{f}\"\n",
    "        key = f.split(\"_\")[0]\n",
    "        sdf = pd.read_csv(fullPath)\n",
    "        sdf = sdf.rename(columns={'Unnamed: 0' : 'gene'})\n",
    "        sdf = sdf.set_index('gene')\n",
    "        sdf = sdf.T\n",
    "        sdf.index = sdf.index.str.replace(\"-\", \".\")\n",
    "        spt[key] = sdf\n",
    "    \n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "85f001e2-d2d5-4e8b-bb51-c44a6b20b586",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cdf.shape=(14848, 4)\n",
      "HFD8     6788\n",
      "ND       4072\n",
      "HFD14    3988\n",
      "Name: key, dtype: int64\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spotId</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACATTTCCCGGATT.1</td>\n",
       "      <td>1139.783401</td>\n",
       "      <td>1365.804581</td>\n",
       "      <td>HFD14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACCGGGTAGGTACC.1</td>\n",
       "      <td>468.056902</td>\n",
       "      <td>1045.415505</td>\n",
       "      <td>HFD14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCGTTCGTCCAGG.1</td>\n",
       "      <td>604.524374</td>\n",
       "      <td>1214.452428</td>\n",
       "      <td>HFD14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACCTAAGCAGCCGG.1</td>\n",
       "      <td>1003.708924</td>\n",
       "      <td>1433.743571</td>\n",
       "      <td>HFD14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCTCATGAAGTTG.1</td>\n",
       "      <td>380.369919</td>\n",
       "      <td>960.970730</td>\n",
       "      <td>HFD14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spotId            x            y    key\n",
       "0  AAACATTTCCCGGATT.1  1139.783401  1365.804581  HFD14\n",
       "1  AAACCGGGTAGGTACC.1   468.056902  1045.415505  HFD14\n",
       "2  AAACCGTTCGTCCAGG.1   604.524374  1214.452428  HFD14\n",
       "3  AAACCTAAGCAGCCGG.1  1003.708924  1433.743571  HFD14\n",
       "4  AAACCTCATGAAGTTG.1   380.369919   960.970730  HFD14"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordDir = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/CARDInputs/\"\n",
    "\n",
    "dfList = []\n",
    "\n",
    "for f in os.listdir(coordDir):\n",
    "    if \"coord\" in f:\n",
    "        fullPath = f\"{coordDir}{f}\"\n",
    "        key = f.split(\"_\")[0]\n",
    "        cdf = pd.read_csv(fullPath)\n",
    "        cdf = cdf.rename(columns={'Unnamed: 0' : 'spotId'})\n",
    "        cdf['key'] = key\n",
    "        dfList.append(cdf)\n",
    "\n",
    "cdf = pd.concat(dfList, ignore_index=True)\n",
    "print(f\"{cdf.shape=}\")\n",
    "print(cdf['key'].value_counts())\n",
    "print()\n",
    "cdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c271a45a-6e35-4327-b8ca-12d608f397de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HFD8     3394\n",
      "ND       2036\n",
      "HFD14    1994\n",
      "Name: key, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spotId</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>key</th>\n",
       "      <th>Dendritic cells</th>\n",
       "      <th>Mac4</th>\n",
       "      <th>Stromal cells</th>\n",
       "      <th>NK cells</th>\n",
       "      <th>Mac2</th>\n",
       "      <th>Monocytes</th>\n",
       "      <th>T cells</th>\n",
       "      <th>Mac5</th>\n",
       "      <th>Mac3</th>\n",
       "      <th>Mac1</th>\n",
       "      <th>B cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACATTTCCCGGATT.1</td>\n",
       "      <td>1139.783401</td>\n",
       "      <td>1365.804581</td>\n",
       "      <td>HFD14</td>\n",
       "      <td>3.468007e-02</td>\n",
       "      <td>0.007548</td>\n",
       "      <td>0.409201</td>\n",
       "      <td>0.170860</td>\n",
       "      <td>0.004757</td>\n",
       "      <td>0.041542</td>\n",
       "      <td>0.055275</td>\n",
       "      <td>0.219997</td>\n",
       "      <td>0.010912</td>\n",
       "      <td>0.004188</td>\n",
       "      <td>0.041040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACCGGGTAGGTACC.1</td>\n",
       "      <td>468.056902</td>\n",
       "      <td>1045.415505</td>\n",
       "      <td>HFD14</td>\n",
       "      <td>1.280364e-03</td>\n",
       "      <td>0.088368</td>\n",
       "      <td>0.187514</td>\n",
       "      <td>0.019633</td>\n",
       "      <td>0.028867</td>\n",
       "      <td>0.129969</td>\n",
       "      <td>0.040070</td>\n",
       "      <td>0.450618</td>\n",
       "      <td>0.008903</td>\n",
       "      <td>0.013757</td>\n",
       "      <td>0.031021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCGTTCGTCCAGG.1</td>\n",
       "      <td>604.524374</td>\n",
       "      <td>1214.452428</td>\n",
       "      <td>HFD14</td>\n",
       "      <td>3.752028e-04</td>\n",
       "      <td>0.004256</td>\n",
       "      <td>0.283055</td>\n",
       "      <td>0.187004</td>\n",
       "      <td>0.021560</td>\n",
       "      <td>0.108340</td>\n",
       "      <td>0.044845</td>\n",
       "      <td>0.310869</td>\n",
       "      <td>0.003200</td>\n",
       "      <td>0.004016</td>\n",
       "      <td>0.032480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AAACCTAAGCAGCCGG.1</td>\n",
       "      <td>1003.708924</td>\n",
       "      <td>1433.743571</td>\n",
       "      <td>HFD14</td>\n",
       "      <td>1.198581e-07</td>\n",
       "      <td>0.009084</td>\n",
       "      <td>0.239973</td>\n",
       "      <td>0.043516</td>\n",
       "      <td>0.004948</td>\n",
       "      <td>0.057287</td>\n",
       "      <td>0.052469</td>\n",
       "      <td>0.558332</td>\n",
       "      <td>0.000064</td>\n",
       "      <td>0.023214</td>\n",
       "      <td>0.011113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AAACCTCATGAAGTTG.1</td>\n",
       "      <td>380.369919</td>\n",
       "      <td>960.970730</td>\n",
       "      <td>HFD14</td>\n",
       "      <td>1.377525e-02</td>\n",
       "      <td>0.029749</td>\n",
       "      <td>0.216199</td>\n",
       "      <td>0.122802</td>\n",
       "      <td>0.053188</td>\n",
       "      <td>0.158933</td>\n",
       "      <td>0.041466</td>\n",
       "      <td>0.234411</td>\n",
       "      <td>0.043126</td>\n",
       "      <td>0.037127</td>\n",
       "      <td>0.049224</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               spotId            x            y    key  Dendritic cells  \\\n",
       "0  AAACATTTCCCGGATT.1  1139.783401  1365.804581  HFD14     3.468007e-02   \n",
       "2  AAACCGGGTAGGTACC.1   468.056902  1045.415505  HFD14     1.280364e-03   \n",
       "4  AAACCGTTCGTCCAGG.1   604.524374  1214.452428  HFD14     3.752028e-04   \n",
       "6  AAACCTAAGCAGCCGG.1  1003.708924  1433.743571  HFD14     1.198581e-07   \n",
       "8  AAACCTCATGAAGTTG.1   380.369919   960.970730  HFD14     1.377525e-02   \n",
       "\n",
       "       Mac4  Stromal cells  NK cells      Mac2  Monocytes   T cells      Mac5  \\\n",
       "0  0.007548       0.409201  0.170860  0.004757   0.041542  0.055275  0.219997   \n",
       "2  0.088368       0.187514  0.019633  0.028867   0.129969  0.040070  0.450618   \n",
       "4  0.004256       0.283055  0.187004  0.021560   0.108340  0.044845  0.310869   \n",
       "6  0.009084       0.239973  0.043516  0.004948   0.057287  0.052469  0.558332   \n",
       "8  0.029749       0.216199  0.122802  0.053188   0.158933  0.041466  0.234411   \n",
       "\n",
       "       Mac3      Mac1   B cells  \n",
       "0  0.010912  0.004188  0.041040  \n",
       "2  0.008903  0.013757  0.031021  \n",
       "4  0.003200  0.004016  0.032480  \n",
       "6  0.000064  0.023214  0.011113  \n",
       "8  0.043126  0.037127  0.049224  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"Merge CARD predictions with spatial coordinates\"\"\"\n",
    "df = pd.merge(cdf, df,\n",
    "              how='left', \n",
    "              left_on=['spotId', 'key'],\n",
    "              right_on=['spotId', 'key'])\n",
    "\n",
    "df = df.fillna(0)\n",
    "df = df.drop_duplicates() # very important!\n",
    "print(df['key'].value_counts())\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "52bc84f1-9d58-4ebc-8490-cb48c3203227",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bb8a323-e33c-45fd-adab-9fd42a8d4b60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ND mean: 110.2 median: 91.0\n"
     ]
    }
   ],
   "source": [
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.facecolor'] = \"w\"\n",
    "plt.rcParams['figure.figsize'] = 9, 3\n",
    "\n",
    "allMean = []\n",
    "\n",
    "fig, axs = plt.subplots(1, 3, sharey=True)\n",
    "axs = axs.ravel()\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    sdf = spt[key]\n",
    "    \n",
    "    bdf = sdf.astype(bool)\n",
    "    rowSums = bdf.sum(axis=1)\n",
    "    \n",
    "    allMean += rowSums.to_list()\n",
    "    \n",
    "    print(f\"{key} mean: {rowSums.mean():.1f} median: {rowSums.median():.1f}\")\n",
    "    \n",
    "    axs[i].hist(rowSums, \n",
    "                edgecolor='k', \n",
    "                bins=30)\n",
    "\n",
    "    axs[i].set_xlabel(\"N Genes\")\n",
    "    if i == 0:\n",
    "        axs[i].set_ylabel(\"N Spots\")\n",
    "    else:\n",
    "        axs[i].set_ylabel(\"\")\n",
    "    axs[i].set_title(key)\n",
    "    \n",
    "    axs[i].axvline(x=rowSums.median(), \n",
    "                   c='r', \n",
    "                   ls=\"--\", \n",
    "                   label=f\"Median={int(rowSums.median())}\")\n",
    "    \n",
    "    axs[i].legend(loc='upper right')\n",
    "    \n",
    "    # break\n",
    "    \n",
    "plt.tight_layout()\n",
    "sns.despine()\n",
    "\n",
    "print(np.mean(allMean))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "48fbd525-a290-4f0a-b07a-79c877e2fe31",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"/tmp/ipykernel_3133793/668683560.py\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6872bdee-1593-4539-9823-bed0350b0bb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "lrData = {}\n",
    "\n",
    "for key in keys:\n",
    "    sdf = spt[key]\n",
    "    newCols = {}\n",
    "    \n",
    "    for idx, row in lr.iterrows():\n",
    "        l = row['ligand']\n",
    "        r = row['receptor']\n",
    "        \n",
    "        if l in sdf.columns and r in sdf.columns:\n",
    "            gMean = gmean(sdf[[l, r]], axis=1)\n",
    "            lrName = row['Pair.Name']\n",
    "            newCols[lrName] = gMean\n",
    "    \n",
    "    lrdf = pd.DataFrame(newCols)\n",
    "    lrdf.index = sdf.index\n",
    "    \n",
    "    # remove zero-sum columns \n",
    "    lrdf = lrdf.loc[:, (lrdf.sum(axis=0) != 0)]\n",
    "    print(sdf.shape, lrdf.shape)\n",
    "    \n",
    "    lrData[key] = lrdf\n",
    "\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06ed0779-864a-4cd2-b9b4-14586cbe4967",
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 25\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.facecolor'] = \"w\"\n",
    "plt.rcParams['figure.figsize'] = 12, 4.75\n",
    "\n",
    "fig, axs = plt.subplots(1, 3)\n",
    "axs = axs.ravel()\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    ldf = lrData[key]\n",
    "    ldfBool = ldf.astype(bool) \n",
    "    colSumns = ldfBool.sum(axis=0)\n",
    "    prop = colSumns / len(ldfBool)\n",
    "    prop = pd.DataFrame(prop)\n",
    "    prop = prop.reset_index(drop=False)\n",
    "    prop.columns = ['LR', 'Proportion']\n",
    "    prop = prop.sort_values(by='Proportion', ascending=False)\n",
    "    # print(f\"==== {key} ====\")\n",
    "    # print(prop.head(n))\n",
    "    \n",
    "    sns.barplot(data=prop.head(n),\n",
    "                y='LR', \n",
    "                x='Proportion', \n",
    "                color='grey',\n",
    "                edgecolor='k',\n",
    "                ax=axs[i])\n",
    "    \n",
    "    \n",
    "    axs[i].set_ylabel(\"\")\n",
    "    axs[i].set_xlabel(\"Proportion of Spots\")\n",
    "    axs[i].set_title(key)\n",
    "    \n",
    "plt.tight_layout()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d06f7962-ed89-4634-971b-32fe016e138d",
   "metadata": {},
   "outputs": [],
   "source": [
    "res = []\n",
    "\n",
    "top = 20\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    sdf = spt[key]\n",
    "    ldf = lrData[key]\n",
    "    ldfBool = ldf.astype(bool) \n",
    "    colSums = ldfBool.sum(axis=0)\n",
    "    colSums = colSums * 1000 / len(sdf) # colocalization per 1000 spots\n",
    "    \n",
    "    res.append(colSums)\n",
    "    \n",
    "res = pd.concat(res, axis=1)\n",
    "res.columns = keys\n",
    "res = res.reset_index(drop=False)\n",
    "\n",
    "# res.head()\n",
    "\n",
    "\n",
    "res['HFD8vND'] = res['HFD8'] - res['ND']\n",
    "res['HFD14vHFD8'] = res['HFD14'] - res['HFD8']\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.facecolor'] = \"w\"\n",
    "plt.rcParams['figure.figsize'] = 10, 8.5\n",
    "fig, axs = plt.subplots(2, 2)\n",
    "\n",
    "\n",
    "for i, up in enumerate([False, True]):\n",
    "    for j, diff in enumerate(['HFD8vND', 'HFD14vHFD8']):    \n",
    "        pdf = res.sort_values(by=diff, ascending=up)\n",
    "        pdf = pdf.head(top)\n",
    "        pdf = pd.melt(pdf, id_vars='index', value_vars=keys)\n",
    "        \n",
    "        sns.scatterplot(data=pdf, \n",
    "                          x='value',\n",
    "                          y='index',\n",
    "                          ax=axs[i, j],\n",
    "                          hue='variable',\n",
    "                          size='value',\n",
    "                          sizes=(7, 135),\n",
    "                          legend=False,\n",
    "                          linewidth=1,\n",
    "                          ec='k',\n",
    "                          palette='viridis',\n",
    "                          zorder=2)\n",
    "        \n",
    "        axs[i, j].grid(True, zorder=1)\n",
    "        axs[i, j].set_ylabel(\"\")\n",
    "        axs[i, j].set_xlabel(\"Colocalization per 1k Spots\")\n",
    "        \n",
    "#         print(pdf.head())\n",
    "        \n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e973ae19-d9bd-43ed-934f-f31a11e2dda1",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = 'WikiPathways_2019_Mouse'\n",
    "# db = 'KEGG_2019_Mouse'\n",
    "# db = 'ontology'\n",
    "\n",
    "top = 20\n",
    "alpha = 0.05\n",
    "npathways = 10\n",
    "ngenes = 10\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.facecolor'] = \"w\"\n",
    "plt.rcParams['figure.figsize'] = 14, 8\n",
    "fig, axs = plt.subplots(2, 2)\n",
    "\n",
    "for i, up in enumerate([False, True]):\n",
    "    for j, diff in enumerate(['HFD8vND', 'HFD14vHFD8']):    \n",
    "        pdf = res.sort_values(by=diff, ascending=up)\n",
    "        pdf = pdf.head(top)\n",
    "        \n",
    "        genes = pdf['index'].apply(lambda x: np.array(x.split(\"-\")))\n",
    "        genes = list(set([g for row in genes for g in row]))\n",
    "        \n",
    "        ef = gget.enrichr(genes, database=db)\n",
    "        ef['logp'] = ef['p_val'].apply(lambda x: -np.log10(x))\n",
    "        ef = ef[ef['adj_p_val'] < alpha]\n",
    "        ef['nGene'] = ef['overlapping_genes'].apply(lambda x: len(x))\n",
    "        \n",
    "        \n",
    "        geneCount = Counter([g for row in ef['overlapping_genes'] for g in row])\n",
    "        displayGenes = [x[0] for x in geneCount.most_common(ngenes)]\n",
    "        \n",
    "        ef2 = ef.overlapping_genes.str.join('|').str.get_dummies()\n",
    "        ef2.index = ef['path_name'].apply(lambda x: x.split(\"(\")[0].capitalize())\n",
    "        ef2.index = [ '\\n'.join(wrap(l, 60)) for l in ef2.index ]\n",
    "        ef2 = ef2[displayGenes]\n",
    "        ef2.columns = [str(x).lower().capitalize() for x in ef2.columns]\n",
    "        \n",
    "        sns.heatmap(data=ef2.head(npathways),\n",
    "                    ax=axs[i, j],\n",
    "                    cbar=False,\n",
    "                    linewidth=0.5,\n",
    "                    linecolor='k',\n",
    "                    cmap=plt.cm.get_cmap('Blues', lut=3))\n",
    "        \n",
    "        axs[i, j].set_ylabel(\"\")\n",
    "        axs[i, j].set_xlabel(\"\")\n",
    "        axs[i, j].tick_params(axis='x', labelrotation=90)\n",
    "        \n",
    "        print(ef['path_name'])\n",
    "#         print()\n",
    "    # break\n",
    "    \n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07487840-be8b-4556-97e9-28f32a75f6e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aeaf52d-8009-41db-b922-6825ef04c675",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.facecolor'] = \"w\"\n",
    "plt.rcParams['figure.figsize'] = 4, 10\n",
    "fig, axs = plt.subplots(3, 1)\n",
    "axs = axs.ravel()\n",
    "\n",
    "\n",
    "boxes = {\n",
    "    'ND' : [1000, 1200, 1000, 1200],\n",
    "    'HFD8' :  [1000, 1200, 1000, 1200],\n",
    "    'HFD14' :  [600, 800, 1000, 1200],\n",
    "}\n",
    "\n",
    "pltLabels = ['ND', '8w', '14w']\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    # sdf = spt[key]\n",
    "    ldf = lrData[key]\n",
    "    kdf = df[df['key'] == key]\n",
    "    \n",
    "    pdf = pd.merge(kdf, ldf, \n",
    "                   how='left',\n",
    "                   left_on='spotId',\n",
    "                   right_index=True)\n",
    "    \n",
    "    bbox = boxes[key]\n",
    "    \n",
    "    pdf = pdf[(pdf['x'] > bbox[0]) & (pdf['x'] < bbox[1])]\n",
    "    pdf = pdf[(pdf['y'] > bbox[2]) & (pdf['y'] < bbox[3])]\n",
    "    \n",
    "    \n",
    "    sns.scatterplot(data=pdf,\n",
    "                    x='x',\n",
    "                    y='y',\n",
    "                    hue='Mac5',\n",
    "                    s=100,\n",
    "                    ec='k',\n",
    "                    legend=False,\n",
    "                    # alpha=1,\n",
    "                    ax=axs[i],\n",
    "                    palette=plt.cm.get_cmap('Reds', lut=10))\n",
    "    \n",
    "    axs[i].invert_yaxis()\n",
    "    axs[i].set_xticks([])\n",
    "    axs[i].set_yticks([])\n",
    "    axs[i].set_ylabel(pltLabels[i], rotation=0, labelpad=10)\n",
    "    axs[i].set_xlabel(\"\")\n",
    "    \n",
    "plt.tight_layout()\n",
    "#     \n",
    "#     print(pdf.shape, ldf.shape, kdf.shape)\n",
    "    \n",
    "    \n",
    "#     break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57d6b89c-c5de-4c79-aa3a-fa77979e17d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ldf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bf5cd04-be92-49b1-b7e4-b48806485993",
   "metadata": {},
   "outputs": [],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05a48347-2bda-4080-b9d2-e6cb8b89bc8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def printGSR(ef, top=None, columns=['path_name', 'overlapping_genes']):\n",
    "    \"\"\"A function to print pathway results \"\"\"\n",
    "    n = top\n",
    "    if top is None:\n",
    "        n = len(ef)\n",
    "\n",
    "    ef['overlapping_genes'] = ef['overlapping_genes'].apply(lambda x: \", \".join(x))\n",
    "    printDf = ef[columns].head(n)\n",
    "    print(tabulate.tabulate(printDf, \n",
    "                            headers=columns, \n",
    "                            showindex=False,\n",
    "                            tablefmt='orgtbl'))\n",
    "\n",
    "# 'plain', 'simple', 'grid', 'pipe', 'orgtbl', 'rst', 'mediawiki',\n",
    "# 'latex', 'latex_raw', 'latex_booktabs', 'latex_longtable' and tsv.\n",
    "# Variable `tabulate_formats`contains the list of currently supported formats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7253189b-0bf5-47b8-97bf-2f0fe673062b",
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha = 0.05\n",
    "top = 20\n",
    "db = 'WikiPathways_2019_Mouse'\n",
    "# db = 'KEGG_2019_Mouse'\n",
    "# db = 'ontology'\n",
    "\n",
    "\n",
    "def flatten(l):\n",
    "    return [item for sublist in l for item in sublist]\n",
    "\n",
    "# look at the top ligand-receptors\n",
    "for i, key in enumerate(keys):\n",
    "    ldf = lrData[key]\n",
    "    ldfBool = ldf.astype(bool) \n",
    "    colSumns = ldfBool.sum(axis=0)\n",
    "    prop = colSumns / len(ldfBool)\n",
    "    prop = pd.DataFrame(prop)\n",
    "    prop = prop.reset_index(drop=False)\n",
    "    prop.columns = ['LR', 'Proportion']\n",
    "    prop = prop.sort_values(by='Proportion', ascending=False)\n",
    "    \n",
    "    pairs = prop['LR'].head(top).to_list()\n",
    "    genes = flatten([x.split(\"-\") for x in pairs])\n",
    "    ef = gget.enrichr(genes, database=db)\n",
    "    ef = ef[ef['p_val'] <= alpha]\n",
    "    pd.set_option('display.max_colwidth', 200)\n",
    "    \n",
    "    print(f\"==== {key} ====\")\n",
    "    printGSR(ef, top=10)\n",
    "    print()\n",
    "    print(pairs)\n",
    "    print()\n",
    "    \n",
    "    # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f6f89ce-e4e2-44ce-a170-4d812984af55",
   "metadata": {},
   "outputs": [],
   "source": [
    "queryPairs = [\n",
    "   'Apoe-Lrp1',\n",
    "]\n",
    "\n",
    "cTypes = [\n",
    "    'Mac1',\n",
    "    'Mac2',\n",
    "    'Mac3',\n",
    "    'Mac4',\n",
    "    'Mac5',\n",
    "    'Monocytes',\n",
    "    'T cells',\n",
    "    'NK cells', \n",
    "    'Dendritic cells', \n",
    "    'Stromal cells', \n",
    "    'B cells'\n",
    "]\n",
    "\n",
    "layer = 1\n",
    "\n",
    "for key in keys:\n",
    "    ldf = lrData[key][queryPairs].reset_index(drop=False)\n",
    "    ldf = ldf.rename(columns={'index':'spotId'})\n",
    "    cdf = df[df['key'] == key].reset_index(drop=True)\n",
    "    \n",
    "    cdf['cType'] = cdf[cTypes].apply(lambda x: cTypes[np.argsort(x)[-layer]], axis=1)\n",
    "    \n",
    "    tdf = pd.merge(cdf, ldf, \n",
    "                   how='left', \n",
    "                   left_on='spotId',\n",
    "                   right_on='spotId')\n",
    "    \n",
    "    print(f\"============= {key} =============\")\n",
    "    for q in queryPairs:\n",
    "        print(f\"==== {q} ====\")\n",
    "        qdf = tdf[tdf[q] > 0]\n",
    "        print(qdf['cType'].value_counts(normalize=True), '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c151e7c0-7be2-4666-a115-e075d71a57fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" neighboorhood around an LR pair \"\"\"\n",
    "\n",
    "k = 7\n",
    "\n",
    "queryPairs = [\n",
    "    'Apoe-Lrp1',\n",
    "]\n",
    "\n",
    "\n",
    "for key in keys:\n",
    "    ldf = lrData[key][queryPairs].reset_index(drop=False)\n",
    "    ldf = ldf.rename(columns={'index':'spotId'})\n",
    "    cdf = df[df['key'] == key].reset_index(drop=True)\n",
    "    \n",
    "    X = cdf[['x', 'y']]\n",
    "    tree = BallTree(X, leaf_size=10, metric='euclidean')\n",
    "    \n",
    "    globalProp = pd.DataFrame(cdf[sorted(cTypes)].mean(axis=0)).reset_index()\n",
    "    globalProp.columns = ['Cell Type', 'GlobalProp']\n",
    "    \n",
    "    print(f\"============= {key} =============\")\n",
    "    for q in queryPairs:\n",
    "        print(f\"==== {q} ====\")\n",
    "        mask = ldf[q] > 0\n",
    "        print(f\"{mask.sum()} ({mask.sum() / len(mask):.3f}) spots have {q} > 0\")\n",
    "        \n",
    "        spotIds = ldf[mask]['spotId'].to_list()\n",
    "        mask = cdf['spotId'].isin(spotIds)\n",
    "        gf = cdf[mask]\n",
    "        \n",
    "        dist, ind = tree.query(X[mask], k=k)\n",
    "        \n",
    "        indx = np.unique(np.ravel(ind))\n",
    "        locs = cdf[cdf.index.isin(indx)]\n",
    "        \n",
    "        print(f\"{len(locs)} ({len(locs) / len(mask):.3f}) spots are within {k} neighbors\")\n",
    "        locs = pd.DataFrame(locs[sorted(cTypes)].mean(axis=0)).reset_index()\n",
    "        locs.columns = ['Cell Type', 'QueryProp']\n",
    "        locs['GlobalProp'] = globalProp['GlobalProp']\n",
    "        locs['diff'] = locs['QueryProp'] - locs['GlobalProp']\n",
    "        \n",
    "        locs = locs.sort_values(by='QueryProp', ascending=False)        \n",
    "        print(locs)\n",
    "\n",
    "        print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "198e0e24-74cd-4717-a4e8-85abefbd99ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e954f419-6f13-449c-9a58-adac4fe4fd97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get genes at colocalization spots\n",
    "\n",
    "query1 = 'Mac4' \n",
    "query2 = 'Stromal cells'\n",
    "alpha = 0.05\n",
    "t = 0.1\n",
    "top = 30\n",
    "# db = 'WikiPathways_2019_Mouse'\n",
    "db = 'KEGG_2019_Mouse'\n",
    "# db = 'ontology'\n",
    "\n",
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "# keys = ['HFD8']\n",
    "\n",
    "for key in keys:\n",
    "    sdf = spt[key]\n",
    "    ldf = lrData[key]\n",
    "    cdf = df[df['key'] == key].reset_index(drop=True)\n",
    "    \n",
    "    # get spots where the proportion of both types\n",
    "    # is greater than the threshold\n",
    "    q1 = np.where(cdf[query1] > t, 1, 0)\n",
    "    q2 = np.where(cdf[query2] > t, 1, 0)\n",
    "    mask = (q1 == q2) & (q1 == 1)\n",
    "    coloc = cdf[mask]\n",
    "    \n",
    "    print(f\"{key} {query1} and {query2} colocalized in {coloc.shape[0] / len(cdf):.3f} of spots\")\n",
    "          \n",
    "    \"\"\" query the LR pairs \"\"\"\n",
    "    lrList = ldf.columns    \n",
    "    coloc = pd.merge(coloc, \n",
    "                     ldf,\n",
    "                     left_on='spotId',\n",
    "                     right_index=True,\n",
    "                     how='left')\n",
    "\n",
    "    # get the colunm sums\n",
    "    exp = pd.DataFrame(coloc[lrList].mean(axis=0))\n",
    "    exp = exp.reset_index()\n",
    "    exp.columns = ['LR', 'sumExp']\n",
    "    exp = exp[exp['sumExp'] > 0]\n",
    "    exp = exp.sort_values(by='sumExp', ascending=False)\n",
    "    pairs = exp['LR'].head(top).to_list()\n",
    "    genes = flatten([x.split(\"-\") for x in pairs])\n",
    "    \n",
    "    if len(genes) > 0:\n",
    "        ef = gget.enrichr(genes, database=db)\n",
    "        ef = ef[ef['p_val'] <= alpha]\n",
    "        pd.set_option('display.max_colwidth', 100)\n",
    "    \n",
    "        print(f\"============ {key} LR Analysis ============\")\n",
    "        printGSR(ef, top=top)\n",
    "        print()\n",
    "        print(pairs)\n",
    "        print()\n",
    "    \n",
    "    \"\"\"query the gene expression\"\"\"\n",
    "    coloc = cdf[mask]\n",
    "    geneList = sdf.columns    \n",
    "    coloc = pd.merge(coloc, \n",
    "                     sdf,\n",
    "                     left_on='spotId',\n",
    "                     right_index=True,\n",
    "                     how='left')\n",
    "    \n",
    "    exp = pd.DataFrame(coloc[geneList].mean(axis=0))\n",
    "    exp = exp.reset_index()\n",
    "    exp.columns = ['gene', 'sumExp']\n",
    "    \n",
    "    exp = exp[exp['sumExp'] > 0]\n",
    "    exp = exp.sort_values(by='sumExp', ascending=False)\n",
    "    pairs = exp['gene'].head(top).to_list()\n",
    "    genes = flatten([x.split(\"-\") for x in pairs])\n",
    "    \n",
    "    if len(genes) > 0:\n",
    "        ef = gget.enrichr(genes, database=db)\n",
    "        ef = ef[ef['p_val'] <= alpha]\n",
    "        pd.set_option('display.max_colwidth', 100)\n",
    "\n",
    "        print(f\"============ {key} Gene Analysis ============\")\n",
    "        printGSR(ef, top=top)\n",
    "        print()\n",
    "        print(pairs)\n",
    "        print()\n",
    "    # break\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fce5c132-0d26-41f1-a5c7-559039b15282",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01fe905c-1860-446b-87d2-6044596a92c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" check a type of interaction around a neighborhood\"\"\"\n",
    "\n",
    "query = 'Mac5'\n",
    "alpha = 0.05\n",
    "t = 0.01\n",
    "k = 6\n",
    "top = 5\n",
    "# db = 'WikiPathways_2019_Mouse'\n",
    "db = 'KEGG_2019_Mouse'\n",
    "# db = 'ontology'\n",
    "\n",
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "for key in keys:\n",
    "    sdf = spt[key]\n",
    "    ldf = lrData[key]\n",
    "    cdf = df[df['key'] == key].reset_index(drop=True)\n",
    "    \n",
    "    mask = cdf[query] > t\n",
    "    \n",
    "    print(f\"{mask.sum() / len(mask):.3f}% {key} spots are {query} with t > {t}\")\n",
    "    \n",
    "    # get all the neighbors of the spots meeting the criteria\n",
    "    X = cdf[['x', 'y']]\n",
    "    tree = BallTree(X, leaf_size=10, metric='euclidean')\n",
    "    dist, ind = tree.query(X[mask], k=k)\n",
    "    \n",
    "    indx = np.unique(np.ravel(ind))\n",
    "    locs = cdf[cdf.index.isin(indx)]\n",
    "    \n",
    "    print(f\"{len(indx) / len(mask):.3f}% are within {k} neigherbors of {query} spots\")\n",
    "    \n",
    "    \"\"\" query the LR pairs \"\"\"\n",
    "    lrList = ldf.columns    \n",
    "    locs = pd.merge(locs, \n",
    "                     ldf,\n",
    "                     left_on='spotId',\n",
    "                     right_index=True,\n",
    "                     how='left')\n",
    "\n",
    "    # get the colunm sums\n",
    "    exp = pd.DataFrame(locs[lrList].mean(axis=0))\n",
    "    exp = exp.reset_index()\n",
    "    exp.columns = ['LR', 'sumExp']\n",
    "    exp = exp[exp['sumExp'] > 0]\n",
    "    exp = exp.sort_values(by='sumExp', ascending=False)\n",
    "    pairs = exp['LR'].head(top).to_list()\n",
    "    genes = flatten([x.split(\"-\") for x in pairs])\n",
    "    \n",
    "    if len(genes) > 0:\n",
    "        ef = gget.enrichr(genes, database=db)\n",
    "        ef = ef[ef['p_val'] <= alpha]\n",
    "        pd.set_option('display.max_colwidth', 100)\n",
    "    \n",
    "        print(f\"============ {key} LR Analysis ============\")\n",
    "        printGSR(ef, top=top)\n",
    "        print()\n",
    "        print(pairs)\n",
    "        print()\n",
    "    \n",
    "    \"\"\"query the gene expression\"\"\"\n",
    "    locs = cdf[mask]\n",
    "    geneList = sdf.columns    \n",
    "    locs = pd.merge(locs, \n",
    "                     sdf,\n",
    "                     left_on='spotId',\n",
    "                     right_index=True,\n",
    "                     how='left')\n",
    "    \n",
    "    exp = pd.DataFrame(locs[geneList].mean(axis=0))\n",
    "    exp = exp.reset_index()\n",
    "    exp.columns = ['gene', 'sumExp']\n",
    "    \n",
    "    exp = exp[exp['sumExp'] > 0]\n",
    "    exp = exp.sort_values(by='sumExp', ascending=False)\n",
    "    pairs = exp['gene'].head(top).to_list()\n",
    "    genes = flatten([x.split(\"-\") for x in pairs])\n",
    "    \n",
    "    if len(genes) > 0:\n",
    "        ef = gget.enrichr(genes, database=db)\n",
    "        ef = ef[ef['p_val'] <= alpha]\n",
    "        pd.set_option('display.max_colwidth', 100)\n",
    "\n",
    "        print(f\"============ {key} Gene Analysis ============\")\n",
    "        printGSR(ef, top=top)\n",
    "        print()\n",
    "        print(pairs)\n",
    "        print()\n",
    "    # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb467cae-6c6a-4a33-b33c-0da09d63fd87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bd4c3c7-de97-4f6a-a040-4acb585cf8cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" check a type of interaction around a neighborhood\"\"\"\n",
    "\n",
    "query = ['LGALS3']\n",
    "alpha = 0.05\n",
    "k = 6\n",
    "top = 10\n",
    "# db = 'WikiPathways_2019_Mouse'\n",
    "db = 'KEGG_2019_Mouse'\n",
    "# db = 'ontology'\n",
    "\n",
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "for key in keys:\n",
    "    sdf = spt[key]\n",
    "    ldf = lrData[key]\n",
    "    cdf = df[df['key'] == key].reset_index(drop=True)\n",
    "    \n",
    "    queryF = sdf[query].astype(bool).sum(axis=1)\n",
    "    mask = queryF == len(query)\n",
    "    \n",
    "    print(f\"{mask.sum() / len(mask):.3f}% {key} spots have {query}\")\n",
    "    \n",
    "    if mask.sum() == 0:\n",
    "        print(\"NO SPOTS FOUND!!!\")\n",
    "        continue\n",
    "    \n",
    "    # get all the neighbors of the spots meeting the criteria\n",
    "    X = cdf[['x', 'y']]\n",
    "    \n",
    "    treeMask = cdf['spotId'].isin(sdf.index[mask])\n",
    "    \n",
    "    tree = BallTree(X, leaf_size=10, metric='euclidean')\n",
    "    dist, ind = tree.query(X[treeMask], k=k)\n",
    "    \n",
    "    indx = np.unique(np.ravel(ind))\n",
    "    locs = cdf[cdf.index.isin(indx)]\n",
    "    \n",
    "    print(f\"{len(indx) / len(mask):.3f}% are within {k} neigherbors of {query} spots\")\n",
    "    \n",
    "    \"\"\"get the tissue composition at these spots \"\"\"\n",
    "    props = pd.DataFrame(locs[cTypes].mean(axis=0)).reset_index(drop=False)\n",
    "    props.columns = ['Cell Type', 'Proportion']\n",
    "    props = props.sort_values(by='Proportion', ascending=False)\n",
    "    print(tabulate.tabulate(props, \n",
    "                            headers=props.columns,\n",
    "                            showindex=False,\n",
    "                            tablefmt='orgtbl'))\n",
    "\n",
    "    \"\"\" query the LR pairs \"\"\"\n",
    "    lrList = ldf.columns    \n",
    "    locs = pd.merge(locs, \n",
    "                     ldf,\n",
    "                     left_on='spotId',\n",
    "                     right_index=True,\n",
    "                     how='left')\n",
    "\n",
    "    # get the colunm sums\n",
    "    exp = pd.DataFrame(locs[lrList].mean(axis=0))\n",
    "    exp = exp.reset_index()\n",
    "    exp.columns = ['LR', 'sumExp']\n",
    "    exp = exp[exp['sumExp'] > 0]\n",
    "    exp = exp.sort_values(by='sumExp', ascending=False)\n",
    "    pairs = exp['LR'].head(top).to_list()\n",
    "    genes = flatten([x.split(\"-\") for x in pairs])\n",
    "    \n",
    "    if len(genes) > 0:\n",
    "        ef = gget.enrichr(genes, database=db)\n",
    "        ef = ef[ef['p_val'] <= alpha]\n",
    "        pd.set_option('display.max_colwidth', 100)\n",
    "    \n",
    "        print(f\"============ {key} LR Analysis ============\")\n",
    "        printGSR(ef, top=top)\n",
    "        print()\n",
    "        print(pairs)\n",
    "        print()\n",
    "    \n",
    "    \"\"\"query the gene expression\"\"\"\n",
    "\n",
    "    geneList = sdf.columns    \n",
    "    locs = pd.merge(locs, \n",
    "                     sdf,\n",
    "                     left_on='spotId',\n",
    "                     right_index=True,\n",
    "                     how='left')\n",
    "    \n",
    "    exp = pd.DataFrame(locs[geneList].mean(axis=0))\n",
    "    exp = exp.reset_index()\n",
    "    exp.columns = ['gene', 'sumExp']\n",
    "    \n",
    "    exp = exp[exp['sumExp'] > 0]\n",
    "    exp = exp.sort_values(by='sumExp', ascending=False)\n",
    "    pairs = exp['gene'].head(top).to_list()\n",
    "    genes = flatten([x.split(\"-\") for x in pairs])\n",
    "    \n",
    "    if len(genes) > 0:\n",
    "        ef = gget.enrichr(genes, database=db)\n",
    "        ef = ef[ef['p_val'] <= alpha]\n",
    "        pd.set_option('display.max_colwidth', 100)\n",
    "\n",
    "        print(f\"============ {key} Gene Analysis ============\")\n",
    "        printGSR(ef, top=top)\n",
    "        print()\n",
    "        print(pairs)\n",
    "        print()\n",
    "    # break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "faae5641-449d-4841-9ee3-4c96ad22fce3",
   "metadata": {},
   "outputs": [],
   "source": [
    "4072/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71bdf177-8a6e-41b4-847b-a008c7c5cdbc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
