{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "97cdc744-8133-4fdc-b0b5-edeb92d9b885",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'utils' from '/home/cstansbu/git_repositories/spatial_transcriptomics/clustering/spatial/../utils.py'>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "import scipy\n",
    "import scipy.stats\n",
    "import matplotlib\n",
    "from scipy import interpolate\n",
    "import os\n",
    "import itertools\n",
    "from textwrap import wrap\n",
    "from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar\n",
    "from sklearn.preprocessing import MultiLabelBinarizer\n",
    "import matplotlib.font_manager as fm\n",
    "import scipy.stats\n",
    "import networkx as nx\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn import metrics\n",
    "from mpl_toolkits.axes_grid1.axes_divider import make_axes_locatable\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "from sklearn.neighbors import radius_neighbors_graph\n",
    "import matplotlib.patches as mpatches\n",
    "import sklearn.neighbors\n",
    "from scipy.spatial import distance\n",
    "from skimage.filters import unsharp_mask\n",
    "from skimage.color import rgb2gray\n",
    "from mpl_toolkits.axes_grid1 import ImageGrid\n",
    "from collections import Counter\n",
    "import skimage\n",
    "from Bio import SeqIO\n",
    "from Bio.KEGG import REST\n",
    "from Bio.KEGG.KGML import KGML_parser\n",
    "import io\n",
    "import numpy as np\n",
    "import gget\n",
    "import matplotlib.pyplot as plt\n",
    "from skimage import data, segmentation, feature, future\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from functools import partial\n",
    "import sys\n",
    "\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "sys.path.append(\"../\")\n",
    "import utils as ut\n",
    "from importlib import reload\n",
    "reload(ut)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4f88e10b-9273-46f9-8cf4-079b3cf91ffa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   spot_diameter_fullres  tissue_hires_scalef  fiducial_diameter_fullres   \n",
      "0             257.195160             0.050684                 415.469120  \\\n",
      "1             257.368127             0.051089                 415.748513   \n",
      "2             257.404844             0.049124                 415.807825   \n",
      "\n",
      "   tissue_lowres_scalef    key  \n",
      "0              0.015205     ND  \n",
      "1              0.015327   HFD8  \n",
      "2              0.014737  HFD14  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "6.016495945974647"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# upload the scale factors\n",
    "\n",
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "scales = []\n",
    "\n",
    "for key in keys:\n",
    "    path = f\"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/SPT/{key}/outs/spatial/scalefactors_json.json\"\n",
    "    \n",
    "    tmp = pd.read_json(path, lines=True)\n",
    "    tmp['key'] = key\n",
    "    # tmp['lowres_spot_diameter_pixels'] = tmp['spot_diameter_fullres'] * tmp['tissue_lowres_scalef']\n",
    "    # tmp['lowres_spot_diameter_units'] = tmp['lowres_spot_diameter_pixels'] * 65\n",
    "    \n",
    "    scales.append(tmp)\n",
    "    \n",
    "scales = pd.concat(scales, ignore_index=True)\n",
    "print(scales)\n",
    "# print(scales[['key', 'spot_diameter_fullres', 'lowres_spot_diameter_pixels']])\n",
    "\n",
    "\n",
    "def getscalebar(key, length, scales=scales):\n",
    "    \"\"\"Length is physical units \"\"\"\n",
    "    lookup = scales.loc[scales['key'] == key]\n",
    "    pixels = lookup['spot_diameter_fullres'].values[0]\n",
    "    scaleFactor = lookup['tissue_lowres_scalef'].values[0]\n",
    "    units = 65 # mircons\n",
    "    \n",
    "    fullResPixels = (pixels * length) / units\n",
    "    lowResPixels = fullResPixels * scaleFactor\n",
    "    \n",
    "    return lowResPixels\n",
    "    \n",
    "getscalebar('ND', length=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dee239e4-bb81-497d-bd6b-c3cbac8503db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HFD14 (2000, 1628)\n",
      "ND (2000, 1681)\n",
      "HFD8 (2000, 1692)\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "imDir = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/images/lowres/\"\n",
    "\n",
    "\n",
    "lowres = {}\n",
    "\n",
    "for f in os.listdir(imDir):\n",
    "    if f.endswith(\"png\"):\n",
    "        fullpath = f\"{imDir}{f}\"\n",
    "        img = skimage.io.imread(fullpath, as_gray=True)\n",
    "        key = f.replace(\".png\", \"\")\n",
    "        print(key, img.shape)\n",
    "        lowres[key] = img\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "322ed2b3-1d8f-479d-b7f0-6767e05007a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2000, 8)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pair.Name</th>\n",
       "      <th>Ligand.Mouse</th>\n",
       "      <th>Receptor.Mouse</th>\n",
       "      <th>Source</th>\n",
       "      <th>ManualAnnotation</th>\n",
       "      <th>Ligand.CC</th>\n",
       "      <th>Ligand.GO</th>\n",
       "      <th>Reference (PMID / KEGG)</th>\n",
       "      <th>ligand</th>\n",
       "      <th>receptor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A2m-Lrp1</td>\n",
       "      <td>A2m</td>\n",
       "      <td>Lrp1</td>\n",
       "      <td>Ramilowski</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Secreted</td>\n",
       "      <td>Other</td>\n",
       "      <td>1702392,10652313, 12194978</td>\n",
       "      <td>A2M</td>\n",
       "      <td>LRP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Adgre5-Cd55</td>\n",
       "      <td>Adgre5</td>\n",
       "      <td>Cd55</td>\n",
       "      <td>Baccin</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Membrane</td>\n",
       "      <td>Other</td>\n",
       "      <td>23447688</td>\n",
       "      <td>ADGRE5</td>\n",
       "      <td>CD55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Adipoq-Adipor1</td>\n",
       "      <td>Adipoq</td>\n",
       "      <td>Adipor1</td>\n",
       "      <td>Ramilowski</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Secreted</td>\n",
       "      <td>Other</td>\n",
       "      <td>12802337</td>\n",
       "      <td>ADIPOQ</td>\n",
       "      <td>ADIPOR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Adipoq-Adipor2</td>\n",
       "      <td>Adipoq</td>\n",
       "      <td>Adipor2</td>\n",
       "      <td>Ramilowski</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Secreted</td>\n",
       "      <td>Other</td>\n",
       "      <td>12802337, 12802330</td>\n",
       "      <td>ADIPOQ</td>\n",
       "      <td>ADIPOR2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Adm-Calcrl&amp;Ramp2</td>\n",
       "      <td>Adm</td>\n",
       "      <td>Calcrl&amp;Ramp2</td>\n",
       "      <td>Baccin</td>\n",
       "      <td>Correct</td>\n",
       "      <td>Secreted</td>\n",
       "      <td>Other</td>\n",
       "      <td>10342881</td>\n",
       "      <td>ADM</td>\n",
       "      <td>CALCRL&amp;RAMP2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Pair.Name Ligand.Mouse Receptor.Mouse      Source ManualAnnotation   \n",
       "0          A2m-Lrp1          A2m           Lrp1  Ramilowski          Correct  \\\n",
       "1       Adgre5-Cd55       Adgre5           Cd55      Baccin          Correct   \n",
       "2    Adipoq-Adipor1       Adipoq        Adipor1  Ramilowski          Correct   \n",
       "3    Adipoq-Adipor2       Adipoq        Adipor2  Ramilowski          Correct   \n",
       "4  Adm-Calcrl&Ramp2          Adm   Calcrl&Ramp2      Baccin          Correct   \n",
       "\n",
       "  Ligand.CC Ligand.GO     Reference (PMID / KEGG)  ligand      receptor  \n",
       "0  Secreted     Other  1702392,10652313, 12194978     A2M          LRP1  \n",
       "1  Membrane     Other                    23447688  ADGRE5          CD55  \n",
       "2  Secreted     Other                    12802337  ADIPOQ       ADIPOR1  \n",
       "3  Secreted     Other          12802337, 12802330  ADIPOQ       ADIPOR2  \n",
       "4  Secreted     Other                    10342881     ADM  CALCRL&RAMP2  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/LR/LewisLabUSCS/Mouse/Mouse-2020-Baccin-LR-pairs.xlsx\"\n",
    "\n",
    "lr = pd.read_excel(fPath, engine=\"openpyxl\")\n",
    "print(lr.shape)\n",
    "\n",
    "lr['ligand'] = lr['Ligand.Mouse'].str.upper()\n",
    "lr['receptor'] = lr['Receptor.Mouse'].str.upper()\n",
    "lr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "04db6e8c-433e-48fc-b4df-6d616d563a87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "df.shape=(7418, 13)\n",
      "key\n",
      "HFD8     3391\n",
      "ND       2034\n",
      "HFD14    1993\n",
      "Name: count, dtype: int64\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spotId</th>\n",
       "      <th>B cells</th>\n",
       "      <th>Mac1</th>\n",
       "      <th>Mac3</th>\n",
       "      <th>Monocytes</th>\n",
       "      <th>T cells</th>\n",
       "      <th>NK cells</th>\n",
       "      <th>Stromal cells</th>\n",
       "      <th>Dendritic cells</th>\n",
       "      <th>Mac4</th>\n",
       "      <th>Mac2</th>\n",
       "      <th>Mac5</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACAAGTATCTCCCA.1_ND</td>\n",
       "      <td>0.018691</td>\n",
       "      <td>0.009410</td>\n",
       "      <td>0.010737</td>\n",
       "      <td>0.070782</td>\n",
       "      <td>0.040298</td>\n",
       "      <td>0.077389</td>\n",
       "      <td>0.645332</td>\n",
       "      <td>0.015281</td>\n",
       "      <td>0.045516</td>\n",
       "      <td>0.039354</td>\n",
       "      <td>0.027211</td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACAGCTTTCAGAAG.1_ND</td>\n",
       "      <td>0.008378</td>\n",
       "      <td>0.005700</td>\n",
       "      <td>0.038038</td>\n",
       "      <td>0.089932</td>\n",
       "      <td>0.048618</td>\n",
       "      <td>0.103983</td>\n",
       "      <td>0.575021</td>\n",
       "      <td>0.046758</td>\n",
       "      <td>0.020027</td>\n",
       "      <td>0.035451</td>\n",
       "      <td>0.028094</td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACAGGGTCTATATT.1_ND</td>\n",
       "      <td>0.024325</td>\n",
       "      <td>0.010761</td>\n",
       "      <td>0.024323</td>\n",
       "      <td>0.117606</td>\n",
       "      <td>0.072993</td>\n",
       "      <td>0.121093</td>\n",
       "      <td>0.478383</td>\n",
       "      <td>0.026608</td>\n",
       "      <td>0.042051</td>\n",
       "      <td>0.048258</td>\n",
       "      <td>0.033598</td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACATTTCCCGGATT.1_ND</td>\n",
       "      <td>0.023002</td>\n",
       "      <td>0.014704</td>\n",
       "      <td>0.044181</td>\n",
       "      <td>0.087019</td>\n",
       "      <td>0.053537</td>\n",
       "      <td>0.083922</td>\n",
       "      <td>0.465958</td>\n",
       "      <td>0.053223</td>\n",
       "      <td>0.087758</td>\n",
       "      <td>0.037452</td>\n",
       "      <td>0.049242</td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACCCGAACGAAATC.1_ND</td>\n",
       "      <td>0.021954</td>\n",
       "      <td>0.003784</td>\n",
       "      <td>0.001240</td>\n",
       "      <td>0.038020</td>\n",
       "      <td>0.026386</td>\n",
       "      <td>0.061364</td>\n",
       "      <td>0.775170</td>\n",
       "      <td>0.003725</td>\n",
       "      <td>0.023106</td>\n",
       "      <td>0.025914</td>\n",
       "      <td>0.019336</td>\n",
       "      <td>ND</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  spotId   B cells      Mac1      Mac3  Monocytes   T cells   \n",
       "0  AAACAAGTATCTCCCA.1_ND  0.018691  0.009410  0.010737   0.070782  0.040298  \\\n",
       "1  AAACAGCTTTCAGAAG.1_ND  0.008378  0.005700  0.038038   0.089932  0.048618   \n",
       "2  AAACAGGGTCTATATT.1_ND  0.024325  0.010761  0.024323   0.117606  0.072993   \n",
       "3  AAACATTTCCCGGATT.1_ND  0.023002  0.014704  0.044181   0.087019  0.053537   \n",
       "4  AAACCCGAACGAAATC.1_ND  0.021954  0.003784  0.001240   0.038020  0.026386   \n",
       "\n",
       "   NK cells  Stromal cells  Dendritic cells      Mac4      Mac2      Mac5 key  \n",
       "0  0.077389       0.645332         0.015281  0.045516  0.039354  0.027211  ND  \n",
       "1  0.103983       0.575021         0.046758  0.020027  0.035451  0.028094  ND  \n",
       "2  0.121093       0.478383         0.026608  0.042051  0.048258  0.033598  ND  \n",
       "3  0.083922       0.465958         0.053223  0.087758  0.037452  0.049242  ND  \n",
       "4  0.061364       0.775170         0.003725  0.023106  0.025914  0.019336  ND  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cardOutDir = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/CARDOuputs/\"\n",
    "\n",
    "dfList = []\n",
    "\n",
    "for f in os.listdir(cardOutDir):\n",
    "    if 'global' in f:\n",
    "        fullPath = f\"{cardOutDir}{f}\"\n",
    "        key = f.split(\"_\")[0]\n",
    "        df = pd.read_csv(fullPath)\n",
    "        df = df.rename(columns={'Unnamed: 0' : 'spotId'})\n",
    "        df['key'] = key\n",
    "        dfList.append(df)\n",
    "\n",
    "df = pd.concat(dfList, ignore_index=True)\n",
    "print(f\"{df.shape=}\")\n",
    "print(df['key'].value_counts())\n",
    "print()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "43a2135b-ad95-48f3-8ee0-15b1d7ce62f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HFD8_global_spt.csv (3394, 31053)\n",
      "ND_global_spt.csv (2036, 31053)\n",
      "HFD14_global_spt.csv (1994, 31053)\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "sptDir =  \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/CARDInputs/\"\n",
    "\n",
    "spt = {}\n",
    "\n",
    "for f in os.listdir(sptDir):\n",
    "    if \"spt\" in f and \"global\" in f:\n",
    "        fullPath = f\"{sptDir}{f}\"\n",
    "        key = f.split(\"_\")[0]\n",
    "        sdf = pd.read_csv(fullPath)\n",
    "        sdf = sdf.rename(columns={'Unnamed: 0' : 'gene'})\n",
    "        sdf = sdf.set_index('gene')\n",
    "        sdf = sdf.T\n",
    "        sdf.index = sdf.index.str.replace(\"-\", \".\")\n",
    "        sdf = ut.normalize(sdf, 1e6)\n",
    "        spt[key] = sdf\n",
    "        print(f, sdf.shape)\n",
    "    \n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ff5412f9-21f0-4b5a-8da5-5dea07355dbe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cdf.shape=(7424, 4)\n",
      "key\n",
      "HFD8     3394\n",
      "ND       2036\n",
      "HFD14    1994\n",
      "Name: count, dtype: int64\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spotId</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACAAGTATCTCCCA.1_HFD8</td>\n",
       "      <td>1244.795278</td>\n",
       "      <td>1189.669724</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACACCAATAACTGC.1_HFD8</td>\n",
       "      <td>405.548324</td>\n",
       "      <td>1349.579809</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACAGCTTTCAGAAG.1_HFD8</td>\n",
       "      <td>303.880251</td>\n",
       "      <td>1068.178931</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACAGGGTCTATATT.1_HFD8</td>\n",
       "      <td>344.496391</td>\n",
       "      <td>1138.478061</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACAGTGTTCCTGGG.1_HFD8</td>\n",
       "      <td>648.734268</td>\n",
       "      <td>1595.524585</td>\n",
       "      <td>HFD8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    spotId            x            y   key\n",
       "0  AAACAAGTATCTCCCA.1_HFD8  1244.795278  1189.669724  HFD8\n",
       "1  AAACACCAATAACTGC.1_HFD8   405.548324  1349.579809  HFD8\n",
       "2  AAACAGCTTTCAGAAG.1_HFD8   303.880251  1068.178931  HFD8\n",
       "3  AAACAGGGTCTATATT.1_HFD8   344.496391  1138.478061  HFD8\n",
       "4  AAACAGTGTTCCTGGG.1_HFD8   648.734268  1595.524585  HFD8"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "coordDir = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/CARDInputs/\"\n",
    "\n",
    "dfList = []\n",
    "\n",
    "for f in os.listdir(coordDir):\n",
    "    if \"coord\" in f and \"global\" in f:\n",
    "        fullPath = f\"{coordDir}{f}\"\n",
    "        key = f.split(\"_\")[0]\n",
    "        cdf = pd.read_csv(fullPath)\n",
    "        cdf = cdf.rename(columns={'Unnamed: 0' : 'spotId'})\n",
    "        cdf['key'] = key\n",
    "        dfList.append(cdf)\n",
    "\n",
    "cdf = pd.concat(dfList, ignore_index=True)\n",
    "print(f\"{cdf.shape=}\")\n",
    "print(cdf['key'].value_counts())\n",
    "print()\n",
    "cdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "908f7e5e-f6a8-494e-ab05-9d9d892fdd47",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "key\n",
      "HFD8     3394\n",
      "ND       2036\n",
      "HFD14    1994\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>spotId</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>key</th>\n",
       "      <th>B cells</th>\n",
       "      <th>Mac1</th>\n",
       "      <th>Mac3</th>\n",
       "      <th>Monocytes</th>\n",
       "      <th>T cells</th>\n",
       "      <th>NK cells</th>\n",
       "      <th>Stromal cells</th>\n",
       "      <th>Dendritic cells</th>\n",
       "      <th>Mac4</th>\n",
       "      <th>Mac2</th>\n",
       "      <th>Mac5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAACAAGTATCTCCCA.1_HFD8</td>\n",
       "      <td>1244.795278</td>\n",
       "      <td>1189.669724</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>0.077635</td>\n",
       "      <td>0.022735</td>\n",
       "      <td>0.104811</td>\n",
       "      <td>0.111453</td>\n",
       "      <td>0.125867</td>\n",
       "      <td>0.195064</td>\n",
       "      <td>0.059765</td>\n",
       "      <td>0.111978</td>\n",
       "      <td>0.082445</td>\n",
       "      <td>0.066571</td>\n",
       "      <td>0.041676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAACACCAATAACTGC.1_HFD8</td>\n",
       "      <td>405.548324</td>\n",
       "      <td>1349.579809</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>0.076987</td>\n",
       "      <td>0.041831</td>\n",
       "      <td>0.047343</td>\n",
       "      <td>0.139484</td>\n",
       "      <td>0.082711</td>\n",
       "      <td>0.093085</td>\n",
       "      <td>0.148247</td>\n",
       "      <td>0.071763</td>\n",
       "      <td>0.058744</td>\n",
       "      <td>0.142045</td>\n",
       "      <td>0.097760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAACAGCTTTCAGAAG.1_HFD8</td>\n",
       "      <td>303.880251</td>\n",
       "      <td>1068.178931</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>0.079216</td>\n",
       "      <td>0.043327</td>\n",
       "      <td>0.075868</td>\n",
       "      <td>0.086821</td>\n",
       "      <td>0.086695</td>\n",
       "      <td>0.097747</td>\n",
       "      <td>0.186142</td>\n",
       "      <td>0.090520</td>\n",
       "      <td>0.069670</td>\n",
       "      <td>0.094246</td>\n",
       "      <td>0.089748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAACAGGGTCTATATT.1_HFD8</td>\n",
       "      <td>344.496391</td>\n",
       "      <td>1138.478061</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>0.037682</td>\n",
       "      <td>0.029618</td>\n",
       "      <td>0.093281</td>\n",
       "      <td>0.109521</td>\n",
       "      <td>0.052151</td>\n",
       "      <td>0.061401</td>\n",
       "      <td>0.201332</td>\n",
       "      <td>0.094533</td>\n",
       "      <td>0.143671</td>\n",
       "      <td>0.088173</td>\n",
       "      <td>0.088637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAACAGTGTTCCTGGG.1_HFD8</td>\n",
       "      <td>648.734268</td>\n",
       "      <td>1595.524585</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>0.092691</td>\n",
       "      <td>0.001555</td>\n",
       "      <td>0.002745</td>\n",
       "      <td>0.075636</td>\n",
       "      <td>0.054631</td>\n",
       "      <td>0.052849</td>\n",
       "      <td>0.112912</td>\n",
       "      <td>0.008552</td>\n",
       "      <td>0.258918</td>\n",
       "      <td>0.056771</td>\n",
       "      <td>0.282740</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    spotId            x            y   key   B cells   \n",
       "0  AAACAAGTATCTCCCA.1_HFD8  1244.795278  1189.669724  HFD8  0.077635  \\\n",
       "1  AAACACCAATAACTGC.1_HFD8   405.548324  1349.579809  HFD8  0.076987   \n",
       "2  AAACAGCTTTCAGAAG.1_HFD8   303.880251  1068.178931  HFD8  0.079216   \n",
       "3  AAACAGGGTCTATATT.1_HFD8   344.496391  1138.478061  HFD8  0.037682   \n",
       "4  AAACAGTGTTCCTGGG.1_HFD8   648.734268  1595.524585  HFD8  0.092691   \n",
       "\n",
       "       Mac1      Mac3  Monocytes   T cells  NK cells  Stromal cells   \n",
       "0  0.022735  0.104811   0.111453  0.125867  0.195064       0.059765  \\\n",
       "1  0.041831  0.047343   0.139484  0.082711  0.093085       0.148247   \n",
       "2  0.043327  0.075868   0.086821  0.086695  0.097747       0.186142   \n",
       "3  0.029618  0.093281   0.109521  0.052151  0.061401       0.201332   \n",
       "4  0.001555  0.002745   0.075636  0.054631  0.052849       0.112912   \n",
       "\n",
       "   Dendritic cells      Mac4      Mac2      Mac5  \n",
       "0         0.111978  0.082445  0.066571  0.041676  \n",
       "1         0.071763  0.058744  0.142045  0.097760  \n",
       "2         0.090520  0.069670  0.094246  0.089748  \n",
       "3         0.094533  0.143671  0.088173  0.088637  \n",
       "4         0.008552  0.258918  0.056771  0.282740  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"Merge CARD predictions with spatial coordinates\"\"\"\n",
    "df = pd.merge(cdf, df,\n",
    "              how='left', \n",
    "              left_on=['spotId', 'key'],\n",
    "              right_on=['spotId', 'key'])\n",
    "\n",
    "df = df.fillna(0)\n",
    "df = df.drop_duplicates() # very important!\n",
    "print(df['key'].value_counts())\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6922a0fb-fa5d-40c3-b31a-af3c0cb983f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3066df82-f2e9-4da0-a4d3-a19292ca35f0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a57d8801-7a2e-491a-92f0-cdd25b2fcd9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def closest_node(node, nodes, metric):\n",
    "    closest_index = distance.cdist([node], nodes, metric=metric).argmin()\n",
    "    return closest_index\n",
    "\n",
    "def getSubGraph(kdf, x, y, metric='minkowski', return_center=True):\n",
    "    \"\"\"A function to generate a graph of size n x n\n",
    "    around point (x, y) \"\"\"\n",
    "    \n",
    "    nodes = kdf[['x', 'y']]\n",
    "    centerInd = closest_node([x, y], nodes, metric)\n",
    "    \n",
    "      \n",
    "    # get the n points around that node\n",
    "    nbrs = NearestNeighbors(n_neighbors=n,  \n",
    "                            metric=metric, \n",
    "                            algorithm='ball_tree').fit(nodes)\n",
    "    distances, indices = nbrs.kneighbors(nodes)\n",
    "    \n",
    "    # create the subgraph\n",
    "    windowIndices = indices[centerInd]\n",
    "    mask = kdf.index.isin(windowIndices)\n",
    "    \n",
    "    center = kdf[mask].reset_index(drop=True)\n",
    "    dists = sklearn.metrics.pairwise_distances(center[['x', 'y']], metric=metric)\n",
    "    \n",
    "    X = dists\n",
    "    \n",
    "    # create the graph\n",
    "    nodeMapper = dict(zip(center.index, center['spotId']))\n",
    "    G = nx.from_numpy_array(X)\n",
    "    G = nx.relabel_nodes(G, nodeMapper)\n",
    "    \n",
    "    # set positions\n",
    "    pos = {}\n",
    "    for idx, row in center.iterrows():\n",
    "        pos[row['spotId']] = np.array(row[['x', 'y']])\n",
    "    \n",
    "    G.pos = pos\n",
    "    \n",
    "    if return_center:\n",
    "        return G, center\n",
    "    else:\n",
    "        return G\n",
    "    \n",
    "    \n",
    "def getFullGraph(kdf, metric='minkowski'):\n",
    "    \"\"\"A function to generate a graph of size n x n\n",
    "    around point (x, y) \"\"\"\n",
    "    \n",
    "    nodes = kdf[['x', 'y']]\n",
    "\n",
    "    dists = sklearn.metrics.pairwise_distances(nodes, metric=metric)\n",
    "    np.fill_diagonal(dists, 0) # no self loops!\n",
    "    X = dists\n",
    "    \n",
    "    # create the graph\n",
    "    nodeMapper = dict(zip(kdf.index, kdf['spotId']))\n",
    "    G = nx.from_numpy_array(X)\n",
    "    G = nx.relabel_nodes(G, nodeMapper)\n",
    "    \n",
    "    # set positions\n",
    "    pos = {}\n",
    "    for idx, row in kdf.iterrows():\n",
    "        pos[row['spotId']] = np.array(row[['x', 'y']])\n",
    "    \n",
    "    G.pos = pos\n",
    "    return G\n",
    "    \n",
    "    \n",
    "def getKnn(kdf, k=6, metric='minkowski'):\n",
    "    \"\"\"A function to generate a knn graph \"\"\"\n",
    "    \n",
    "    nodes = kdf[['x', 'y']]\n",
    "      \n",
    "    # get the n points around that node\n",
    "    nbrs = NearestNeighbors(n_neighbors=k+1,  \n",
    "                            metric=metric, \n",
    "                            algorithm='ball_tree').fit(nodes)\n",
    "    distances, indices = nbrs.kneighbors(nodes)\n",
    "    return indices\n",
    "\n",
    "\n",
    "def knnInd2Id(kdf, knnInd):\n",
    "    \"\"\"Translate neighbors to spotIds \"\"\"\n",
    "    \n",
    "    transId = {}\n",
    "    \n",
    "    for i, row in kdf.iterrows():\n",
    "        spotId = row['spotId']\n",
    "        nbrhs = kdf.iloc[knnInd[i]]['spotId'].values\n",
    "        transId[spotId] = nbrhs\n",
    "    return transId\n",
    "    \n",
    "    \n",
    "\n",
    "def blur(sdf, transId):\n",
    "    \"\"\"A function to blur all columns using their neighbors.\n",
    "    Expects spotIds as index \"\"\"\n",
    "    \n",
    "    results  = []\n",
    "    \n",
    "    for spotId, nbhs in transId.items():\n",
    "        \n",
    "        # make sure the spot is present\n",
    "        if spotId in sdf.index:\n",
    "            nbhs = [x for x in nbhs if x in sdf.index]\n",
    "            \n",
    "            t = sdf.loc[nbhs].mean().to_dict()\n",
    "            t['spotId'] = spotId\n",
    "            results.append(t)\n",
    "        \n",
    "    results = pd.DataFrame(results)\n",
    "    results = results.set_index('spotId')\n",
    "    return results\n",
    "\n",
    "\n",
    "def getCentralities(G):\n",
    "    \"\"\" function to compute centralities \"\"\"\n",
    "    cf = pd.DataFrame({'spotId' : G.nodes()})\n",
    "    cf['degree_centrality'] = cf['spotId'].map(nx.degree_centrality(G))\n",
    "    cf['betweenness_centrality'] = cf['spotId'].map(nx.betweenness_centrality(G))\n",
    "    # cf['closeness_centrality'] = cf['spotId'].map(nx.closeness_centrality(G))\n",
    "    # cf['harmonic_centrality'] = cf['spotId'].map(nx.harmonic_centrality(G))\n",
    "    cf['eigenvector_centrality'] = cf['spotId'].map(nx.eigenvector_centrality(G, tol=0.001))\n",
    "    return cf\n",
    "\n",
    "\n",
    "def prune(G, weight, threshold, min_components):\n",
    "    \"\"\"A function to prune disconnected components \"\"\"\n",
    "    H = nx.Graph(((u, v, e) for u,v,e in G.edges(data=True) if e[weight] < threshold))\n",
    "    H.pos = G.pos\n",
    "    \n",
    "    for component in list(nx.connected_components(H)):\n",
    "        if len(component) < min_components:\n",
    "            for node in component:\n",
    "                H.remove_node(node)\n",
    "    return H\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "def removeDisconnectedNodes(G, radius=25, min_edges=3):\n",
    "    \"\"\"A function to remove spots that are not nearby\n",
    "    other spots \"\"\"\n",
    "    H = G.copy()\n",
    "    H.pos = G.pos\n",
    "    \n",
    "    H.remove_edges_from([(n1, n2) for n1, n2, w in H.edges(data=\"weight\") if w > radius])\n",
    "    degree = pd.DataFrame(nx.degree(H), columns=['node', 'degree'])\n",
    "    \n",
    "    mask = (degree['degree'] > min_edges)\n",
    "    dropNodes = degree[~mask]\n",
    "    \n",
    "    # drop the nodes\n",
    "    H.remove_nodes_from(dropNodes['node'].to_list())\n",
    "    \n",
    "    return H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "92d319ba-b04d-4d3f-a2f0-b9c72c663367",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "graph = {}\n",
    "min_components = 3 # remove disconnected nodes\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "NOTE: here we remove edges that are greater distance than the nearest neighbor\n",
    "\"\"\" \n",
    "    \n",
    "for i, key in enumerate(keys):\n",
    "    kdf = df[df['key'] == key].reset_index(drop=True)\n",
    "    G = getFullGraph(kdf)\n",
    "    G = removeDisconnectedNodes(G)\n",
    "    graph[key] = G\n",
    "    \n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7fa041bf-c36b-4f1d-b57d-cc0297d17538",
   "metadata": {},
   "outputs": [],
   "source": [
    "cTypes = [\n",
    "    'Dendritic cells', \n",
    "    'Mac4',\n",
    "    'Stromal cells',\n",
    "    'NK cells', \n",
    "    'Mac2', \n",
    "    'Monocytes',\n",
    "    'T cells',\n",
    "    'Mac5',\n",
    "    'Mac3',\n",
    "    'Mac1',\n",
    "    'B cells',\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c5184209-28d9-4322-b11a-476b4485cc3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done\n"
     ]
    }
   ],
   "source": [
    "\"\"\"Build the graph\"\"\"\n",
    "\n",
    "allGraphs = {}\n",
    "radius = 25\n",
    "\n",
    "for i, key in enumerate(keys):\n",
    "    \n",
    "    kdf = df[df['key'] == key].reset_index(drop=True)\n",
    "    G = graph[key]\n",
    "    \n",
    "    normedCols = {}\n",
    "    for ctype in cTypes:\n",
    "        newCol = f\"{ctype} norm\"\n",
    "        normedCols[newCol] = ctype\n",
    "        kdf[newCol] = kdf[ctype] / kdf[ctype].max()\n",
    "        \n",
    "    node_attr = kdf.set_index('spotId').to_dict('index')\n",
    "    nx.set_node_attributes(G, node_attr)    \n",
    "    \n",
    "    edgelists = {}\n",
    "    for u,v,e in G.edges(data=True):\n",
    "        if e['weight'] < radius:\n",
    "            \n",
    "            for norm, ctype in normedCols.items():\n",
    "                x1 = G.nodes[u][norm]\n",
    "                x2 = G.nodes[v][norm]\n",
    "                \n",
    "                w = scipy.stats.hmean([x1, x2])\n",
    "                edge = (u,v, w)\n",
    "                \n",
    "                if ctype in edgelists.keys():\n",
    "                    edgelists[ctype].append(edge)\n",
    "                else:\n",
    "                    edgelists[ctype] = [edge]\n",
    "                \n",
    "    \"\"\" create the subgraph \"\"\"\n",
    "    H = nx.Graph()\n",
    "    H.add_nodes_from(G.nodes())\n",
    "    \n",
    "    for ctype in cTypes:\n",
    "        edges = edgelists[ctype]\n",
    "        H.add_weighted_edges_from(edges, weight=ctype)\n",
    "    H.pos = G.pos\n",
    "    nx.set_node_attributes(H, node_attr)  \n",
    "    allGraphs[key] = H\n",
    "    \n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bb4a1b87-0f08-48e3-8421-9fbc1f0ca5bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " ND Dendritic cells 0.002\n",
      " ND Mac1 0.007\n",
      " ND Mac2 0.133\n",
      " ND Mac3 0.001\n",
      " ND Mac4 0.083\n",
      " ND Mac5 0.018\n",
      " ND Monocytes 0.153\n",
      " ND T cells 0.106\n",
      " HFD8 Dendritic cells 0.000\n",
      " HFD8 Mac1 0.000\n",
      " HFD8 Mac2 0.000\n",
      " HFD8 Mac3 0.000\n",
      " HFD8 Mac4 0.000\n",
      " HFD8 Mac5 0.000\n",
      " HFD8 Monocytes 0.000\n",
      " HFD8 T cells 0.000\n",
      " HFD14 Dendritic cells 0.033\n",
      " HFD14 Mac1 0.009\n",
      " HFD14 Mac2 0.228\n",
      " HFD14 Mac3 0.040\n",
      " HFD14 Mac4 0.111\n",
      " HFD14 Mac5 0.130\n",
      " HFD14 Monocytes 0.201\n",
      " HFD14 T cells 0.032\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>log2foldchange</th>\n",
       "      <th>key</th>\n",
       "      <th>cellType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XKR4</td>\n",
       "      <td>2.212847</td>\n",
       "      <td>ND</td>\n",
       "      <td>Dendritic cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GM1992</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ND</td>\n",
       "      <td>Dendritic cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GM37381</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ND</td>\n",
       "      <td>Dendritic cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RP1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ND</td>\n",
       "      <td>Dendritic cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SOX17</td>\n",
       "      <td>0.330848</td>\n",
       "      <td>ND</td>\n",
       "      <td>Dendritic cells</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  log2foldchange key         cellType\n",
       "0     XKR4        2.212847  ND  Dendritic cells\n",
       "1   GM1992        0.000000  ND  Dendritic cells\n",
       "2  GM37381        0.000000  ND  Dendritic cells\n",
       "3      RP1        0.000000  ND  Dendritic cells\n",
       "4    SOX17        0.330848  ND  Dendritic cells"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = 0.998 # the quantile threshold for edge existence, closer to 1 allows feweer edges\n",
    "\n",
    "ctypes = [\n",
    "    # 'B cells',\n",
    "    'Dendritic cells',\n",
    "    'Mac1',\n",
    "    'Mac2',\n",
    "    'Mac3',\n",
    "    'Mac4',\n",
    "    'Mac5',\n",
    "    'Monocytes',\n",
    "    # 'NK cells',\n",
    "    # 'Stromal cells',\n",
    "    'T cells'\n",
    "]\n",
    "\n",
    "\n",
    "keys = ['ND', 'HFD8', \"HFD14\"]\n",
    "\n",
    "res = []\n",
    "\n",
    "for key in keys:\n",
    "    sdf = spt[key] # get spatial data\n",
    "    G = allGraphs[key] # get the graph\n",
    "    for i, ct in enumerate(ctypes):\n",
    "\n",
    "        A = nx.to_pandas_adjacency(G, weight=ct)\n",
    "        T = np.quantile(A, q)\n",
    "        print(f\" {key} {ct} {T:.3f}\")\n",
    "\n",
    "        # binarize the matrix\n",
    "        A = pd.DataFrame(np.where(A > T, 1, 0), index=A.index, columns=A.columns)\n",
    "        connected = A[A.sum() > 0].index\n",
    "\n",
    "        mask = sdf.index.isin(connected)\n",
    "        inGroup = sdf[mask].mean()\n",
    "        outGroup = sdf[~mask].mean()\n",
    "\n",
    "        lfc = (inGroup + 1) / (outGroup + 1)\n",
    "        lfc = np.log2(lfc)\n",
    "        lfc = pd.DataFrame(lfc.reset_index(drop=False))\n",
    "        lfc.columns = ['gene', 'log2foldchange']\n",
    "        lfc['key'] = key\n",
    "        lfc['cellType'] = ct\n",
    "\n",
    "        res.append(lfc)\n",
    "\n",
    "res = pd.concat(res)\n",
    "res.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b443e121-1139-4e7d-af61-4a86edbaa042",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "'break' outside loop (668683560.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn[14], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    break\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m 'break' outside loop\n"
     ]
    }
   ],
   "source": [
    "break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "597924e0-e15d-4e4d-849e-c50f88d40107",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue Jun 20 13:20:35 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:20:41 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:20:46 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:20:51 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:20:56 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:03 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:09 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:14 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:19 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:30 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:36 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:42 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:47 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:53 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:21:58 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:22:04 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:22:09 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:22:12 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:22:16 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:22:19 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:22:22 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:22:26 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:22:29 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n",
      "Tue Jun 20 13:22:32 2023 INFO Performing Enichr analysis using database KEGG_2019_Mouse.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(165, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>path_name</th>\n",
       "      <th>overlapping_genes</th>\n",
       "      <th>adj_p_val</th>\n",
       "      <th>key</th>\n",
       "      <th>cellType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Basal transcription factors</td>\n",
       "      <td>[TAF15, CCNH, TAF11, GTF2H1, GTF2E1, GTF2E2, T...</td>\n",
       "      <td>0.042944</td>\n",
       "      <td>ND</td>\n",
       "      <td>Dendritic cells</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Basal transcription factors</td>\n",
       "      <td>[TAF15, CCNH, TAF11, GTF2H1, GTF2E1, GTF2E2, T...</td>\n",
       "      <td>0.005606</td>\n",
       "      <td>ND</td>\n",
       "      <td>Mac1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Nucleotide excision repair</td>\n",
       "      <td>[PCNA, CCNH, CETN2, RPA1, GTF2H1, CUL4A, POLD4...</td>\n",
       "      <td>0.031093</td>\n",
       "      <td>ND</td>\n",
       "      <td>Mac2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>[SEC23A, TUSC3, HSPA4L, RPN1, NGLY1, CUL1, UBE...</td>\n",
       "      <td>0.000812</td>\n",
       "      <td>ND</td>\n",
       "      <td>Mac4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Basal transcription factors</td>\n",
       "      <td>[TAF15, TAF12, CCNH, TAF11, GTF2H1, GTF2H3, GT...</td>\n",
       "      <td>0.000812</td>\n",
       "      <td>ND</td>\n",
       "      <td>Mac4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     path_name   \n",
       "0                  Basal transcription factors  \\\n",
       "0                  Basal transcription factors   \n",
       "0                   Nucleotide excision repair   \n",
       "0  Protein processing in endoplasmic reticulum   \n",
       "1                  Basal transcription factors   \n",
       "\n",
       "                                   overlapping_genes  adj_p_val key   \n",
       "0  [TAF15, CCNH, TAF11, GTF2H1, GTF2E1, GTF2E2, T...   0.042944  ND  \\\n",
       "0  [TAF15, CCNH, TAF11, GTF2H1, GTF2E1, GTF2E2, T...   0.005606  ND   \n",
       "0  [PCNA, CCNH, CETN2, RPA1, GTF2H1, CUL4A, POLD4...   0.031093  ND   \n",
       "0  [SEC23A, TUSC3, HSPA4L, RPN1, NGLY1, CUL1, UBE...   0.000812  ND   \n",
       "1  [TAF15, TAF12, CCNH, TAF11, GTF2H1, GTF2H3, GT...   0.000812  ND   \n",
       "\n",
       "          cellType  \n",
       "0  Dendritic cells  \n",
       "0             Mac1  \n",
       "0             Mac2  \n",
       "0             Mac4  \n",
       "1             Mac4  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lfc = 3\n",
    "alpha = 0.05 # for the enrichments\n",
    "\n",
    "pdf = res.copy()\n",
    "pdf = pdf[pdf['log2foldchange'] > lfc]\n",
    "\n",
    "# db = 'WikiPathways_2019_Mouse'\n",
    "db = 'KEGG_2019_Mouse'\n",
    "# db = 'ontology'\n",
    "\n",
    "edf = []\n",
    "\n",
    "for key in keys:\n",
    "    for i, ct in enumerate(ctypes):\n",
    "        kdf = pdf[(pdf['key'] == key) & (pdf['cellType'] == ct)]\n",
    "\n",
    "        ef = gget.enrichr(kdf['gene'].unique(), database=db)\n",
    "        ef = ef[ef['adj_p_val'] <= alpha]\n",
    "        ef = ef[['path_name', 'overlapping_genes', 'adj_p_val']]\n",
    "        ef['key'] = key\n",
    "        ef['cellType'] = ct\n",
    "        if len(ef) > 0:\n",
    "            edf.append(ef)\n",
    "\n",
    "edf = pd.concat(edf)\n",
    "print(edf.shape)\n",
    "edf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "448e084d-e8dd-4afe-a0f7-064b63fd1fb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "path_name\n",
       "Fanconi anemia pathway                         12\n",
       "Nucleotide excision repair                     10\n",
       "Homologous recombination                       10\n",
       "Autophagy                                       9\n",
       "Basal transcription factors                     9\n",
       "Pyrimidine metabolism                           8\n",
       "Notch signaling pathway                         8\n",
       "Amino sugar and nucleotide sugar metabolism     8\n",
       "DNA replication                                 8\n",
       "MAPK signaling pathway                          6\n",
       "Rap1 signaling pathway                          5\n",
       "Cell cycle                                      3\n",
       "Ubiquitin mediated proteolysis                  2\n",
       "Apoptosis                                       2\n",
       "Protein processing in endoplasmic reticulum     1\n",
       "NF-kappa B signaling pathway                    1\n",
       "Sphingolipid metabolism                         1\n",
       "Mismatch repair                                 1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edf['path_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "43284f90-692a-4557-8340-942926934513",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>path_name</th>\n",
       "      <th>overlapping_genes</th>\n",
       "      <th>adj_p_val</th>\n",
       "      <th>key</th>\n",
       "      <th>cellType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Apoptosis</td>\n",
       "      <td>[PIK3CD, PIK3CB, ACTB, CTSS, ACTG1, LMNB1, BBC...</td>\n",
       "      <td>0.032550</td>\n",
       "      <td>ND</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>[TUSC3, NGLY1, UBE2D1, UBE2D2A, HERPUD1, CASP1...</td>\n",
       "      <td>0.032550</td>\n",
       "      <td>ND</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ubiquitin mediated proteolysis</td>\n",
       "      <td>[UBE3C, UBE2D1, CBLB, UBE2L6, UBE2D2A, UBE2L3,...</td>\n",
       "      <td>0.032550</td>\n",
       "      <td>ND</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Basal transcription factors</td>\n",
       "      <td>[GTF2A1, CCNH, GTF2B, GTF2E1, GTF2E2, TAF5L, G...</td>\n",
       "      <td>0.000359</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Nucleotide excision repair</td>\n",
       "      <td>[CCNH, CETN2, XPA, XPC, POLD3, POLD4, POLD1, P...</td>\n",
       "      <td>0.000359</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DNA replication</td>\n",
       "      <td>[RNASEH2C, PRIM2, FEN1, RNASEH2B, MCM7, PRIM1,...</td>\n",
       "      <td>0.009730</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Fanconi anemia pathway</td>\n",
       "      <td>[MUS81, BLM, FAAP24, BRCA2, BRIP1, SLX4, REV3L...</td>\n",
       "      <td>0.011793</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Autophagy</td>\n",
       "      <td>[RAB7, IRS1, MTMR14, PIK3CD, IRS3, PIK3CB, HMG...</td>\n",
       "      <td>0.011793</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Pyrimidine metabolism</td>\n",
       "      <td>[DTYMK, PNP2, NUDT2, TYMS, NT5C2, NT5C3, NT5C,...</td>\n",
       "      <td>0.014888</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Amino sugar and nucleotide sugar metabolism</td>\n",
       "      <td>[TSTA3, CMAS, GALT, UXS1, HEXB, MPI, HEXA, NPL...</td>\n",
       "      <td>0.020541</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Homologous recombination</td>\n",
       "      <td>[MUS81, BLM, ABRAXAS1, BRCC3, BRCA2, RAD54B, B...</td>\n",
       "      <td>0.042952</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Notch signaling pathway</td>\n",
       "      <td>[NOTCH3, NUMBL, HDAC2, NOTCH1, CTBP2, MAML1, H...</td>\n",
       "      <td>0.042952</td>\n",
       "      <td>HFD8</td>\n",
       "      <td>Mac5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     path_name   \n",
       "0                                    Apoptosis  \\\n",
       "1  Protein processing in endoplasmic reticulum   \n",
       "2               Ubiquitin mediated proteolysis   \n",
       "0                  Basal transcription factors   \n",
       "1                   Nucleotide excision repair   \n",
       "2                              DNA replication   \n",
       "3                       Fanconi anemia pathway   \n",
       "4                                    Autophagy   \n",
       "5                        Pyrimidine metabolism   \n",
       "6  Amino sugar and nucleotide sugar metabolism   \n",
       "7                     Homologous recombination   \n",
       "8                      Notch signaling pathway   \n",
       "\n",
       "                                   overlapping_genes  adj_p_val   key cellType  \n",
       "0  [PIK3CD, PIK3CB, ACTB, CTSS, ACTG1, LMNB1, BBC...   0.032550    ND     Mac5  \n",
       "1  [TUSC3, NGLY1, UBE2D1, UBE2D2A, HERPUD1, CASP1...   0.032550    ND     Mac5  \n",
       "2  [UBE3C, UBE2D1, CBLB, UBE2L6, UBE2D2A, UBE2L3,...   0.032550    ND     Mac5  \n",
       "0  [GTF2A1, CCNH, GTF2B, GTF2E1, GTF2E2, TAF5L, G...   0.000359  HFD8     Mac5  \n",
       "1  [CCNH, CETN2, XPA, XPC, POLD3, POLD4, POLD1, P...   0.000359  HFD8     Mac5  \n",
       "2  [RNASEH2C, PRIM2, FEN1, RNASEH2B, MCM7, PRIM1,...   0.009730  HFD8     Mac5  \n",
       "3  [MUS81, BLM, FAAP24, BRCA2, BRIP1, SLX4, REV3L...   0.011793  HFD8     Mac5  \n",
       "4  [RAB7, IRS1, MTMR14, PIK3CD, IRS3, PIK3CB, HMG...   0.011793  HFD8     Mac5  \n",
       "5  [DTYMK, PNP2, NUDT2, TYMS, NT5C2, NT5C3, NT5C,...   0.014888  HFD8     Mac5  \n",
       "6  [TSTA3, CMAS, GALT, UXS1, HEXB, MPI, HEXA, NPL...   0.020541  HFD8     Mac5  \n",
       "7  [MUS81, BLM, ABRAXAS1, BRCC3, BRCA2, RAD54B, B...   0.042952  HFD8     Mac5  \n",
       "8  [NOTCH3, NUMBL, HDAC2, NOTCH1, CTBP2, MAML1, H...   0.042952  HFD8     Mac5  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = edf[edf['cellType'] == 'Mac5']\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c413156-2b19-4231-988f-5f84492dca4f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
