{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06f3c223-d25d-49da-92ea-ab3c1964b282",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib \n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "from sklearn import metrics\n",
    "from sklearn.neighbors import NearestNeighbors\n",
    "import tabulate\n",
    "import itertools\n",
    "from mpl_toolkits import mplot3d\n",
    "import matplotlib.patches as mpatches\n",
    "import umap\n",
    "from importlib import reload\n",
    "from scipy import stats\n",
    "import matplotlib.image as mpimg\n",
    "import gget\n",
    "\n",
    "# locals\n",
    "import utils as ut\n",
    "reload(ut)\n",
    "\n",
    "sc.settings.verbosity = 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80b471a9-4699-4c6e-8618-7fe3732d2ee2",
   "metadata": {},
   "outputs": [],
   "source": [
    "clusterPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/scanpy/clusters.csv\"\n",
    "cf = pd.read_parquet(clusterPath)\n",
    "print(cf.shape)\n",
    "print(cf['cellType'].unique())\n",
    "cf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "332a9137-6b25-462f-8c7b-45b07a21f575",
   "metadata": {},
   "outputs": [],
   "source": [
    "dirPath = \"/nfs/turbo/umms-indikar/shared/projects/spatial_transcriptomics/data/SPT/\"\n",
    "\n",
    "adata = {}\n",
    "\n",
    "for f in os.listdir(dirPath):\n",
    "    \n",
    "    if not f == \"all\":\n",
    "        key = f\n",
    "        mtxPath = f\"{dirPath}{f}/outs/\"\n",
    "        data = sc.read_visium(mtxPath)\n",
    "        data.var_names_make_unique()\n",
    "        data.var_names = [x.upper() for x in data.var_names]\n",
    "        adata[key] = data\n",
    "    \n",
    "adata.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34fa3a06-42e8-4c3f-a017-7f3dad8689c3",
   "metadata": {},
   "source": [
    "# Filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef5d987b-46ac-4947-beee-15330839ecc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "keys = ['ND', 'HFD8', 'HFD14']\n",
    "\n",
    "min_counts = 10\n",
    "\n",
    "for key in keys:\n",
    "    print(key)\n",
    "    data = adata[key]\n",
    "    sc.pp.filter_cells(data, min_counts=min_counts)\n",
    "    sc.pp.normalize_total(data, target_sum=1e4) # Normalize each cell by total counts over all genes\n",
    "    sc.pp.log1p(data) # Logarithmize data via `X = \\log(X + 1)`,\n",
    "    sc.pp.scale(data) # unit variance and zero mean\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c1985bb-d8d5-4182-b879-557b7b354ce3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69cf6adc-10e4-4257-b086-7a65501b1d94",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75fbfad7-31ae-4c0b-989e-9e416290e59a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def getImage(data):\n",
    "    \"\"\"A function to get image from the spatial data \"\"\"\n",
    "    from scanpy.pl._tools import scatterplots as spt    \n",
    "    sptData = data.uns['spatial']\n",
    "    \n",
    "    libraryId = list(sptData.keys())[0]\n",
    "    sptData = sptData[libraryId]\n",
    "    \n",
    "    # get params\n",
    "    img = sptData['images']['hires']\n",
    "    scale_factor = sptData['scalefactors']['tissue_hires_scalef']\n",
    "    return img, scale_factor\n",
    "\n",
    "\n",
    "trim = {\n",
    "    'ND' : {'xlim' :  [250, 1500], 'ylim' : [1600, 650]},\n",
    "    'HFD8' : {'xlim' :  [190, 1350], 'ylim' : [1680, 400]},\n",
    "    'HFD14' : {'xlim' :  [250, 1400], 'ylim' : [1650, 450]},\n",
    "}\n",
    "\n",
    "print(trim)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40a81ade-c202-4e27-aac0-be91745a318b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctypes = sorted(cf['cellType'].unique())\n",
    "nTypes = len(ctypes)\n",
    "colorList = ut.ncolor(nTypes, cmap='Spectral')\n",
    "colorDict = dict(zip(ctypes, colorList))\n",
    "\n",
    "cf['color'] = cf['cellType'].map(colorDict)\n",
    "cf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b72054fd-455d-4377-81bd-32be430b2127",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 10, 10\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "lft = 1\n",
    "nGene = 30\n",
    "alpha = 0.05\n",
    "q = 20\n",
    "layer = 6 # reverse ordered from 6 to 0 with 6 == 1 first layer\n",
    "\n",
    "for key in keys:\n",
    "    data = adata[key]\n",
    "    # get the image and the scale factors\n",
    "    img, scale_factor = getImage(data)\n",
    "    df = data.to_df()\n",
    "    \n",
    "    # get the image coordinates and scale them\n",
    "    coords = data.obsm['spatial']\n",
    "    x = coords[:, 0]\n",
    "    y = coords[:, 1]\n",
    "    x = x * scale_factor \n",
    "    y = y * scale_factor \n",
    "    \n",
    "    # get marker genes for cell types\n",
    "    rf = cf[cf['key'] == key]\n",
    "    rf = rf[rf['pvals'] <= alpha].reset_index(drop=True)\n",
    "    \n",
    "    rf = rf.groupby(['gene', 'cellType']).agg(\n",
    "        clusterCount = ('clusterId', 'count'),\n",
    "        meanLFC = ('logfoldchanges', 'mean')\n",
    "    ).reset_index(drop=False)\n",
    "    \n",
    "    rf = rf[rf['meanLFC'] > lft].reset_index(drop=True)        \n",
    "    rf = rf.set_index('gene')\n",
    "    rf = rf.groupby(['cellType'])['clusterCount'].nlargest(nGene).reset_index(drop=False)\n",
    "    rf['flag'] = 1.0\n",
    "    \n",
    "    # print(rf.head())\n",
    "    # print(rf['cellType'].value_counts())\n",
    "    \n",
    "    rf = pd.pivot_table(rf, values='flag',\n",
    "                        index='gene', \n",
    "                        columns='cellType')\n",
    "    \n",
    "    rf = rf.fillna(0.0)\n",
    "\n",
    "    dfT = df.T\n",
    "    # binarize the spot data\n",
    "    spots = dfT.columns\n",
    "    \n",
    "    spotDecon = pd.merge(dfT, rf,\n",
    "                         how='left',  \n",
    "                         left_on=dfT.index,\n",
    "                         right_on=rf.index,)\n",
    "    \n",
    "    spotDecon = spotDecon.fillna(0)\n",
    "    \"\"\" Deconvolute the spots \"\"\"\n",
    "    \n",
    "    spotColor = []\n",
    "    spotTypes = []\n",
    "    for spot in spots:\n",
    "        spotThresh = np.percentile(spotDecon[spot], q)\n",
    "        spotVec = np.where(spotDecon[spot] > spotThresh, 1.0, 0.0)\n",
    "        scores = []\n",
    "        for c in ctypes:\n",
    "            \n",
    "            score = metrics.jaccard_score(spotDecon[c], spotVec)\n",
    "            scores.append(score)\n",
    "        \n",
    "        indx = np.argsort(scores)\n",
    "        i = np.nonzero(indx == layer)[0][0]\n",
    "        spotType = ctypes[i]\n",
    "        spotColor.append(colorDict[spotType])\n",
    "        spotTypes.append(spotType)\n",
    "        \n",
    "    print(pd.Series(spotTypes).value_counts())\n",
    "        \n",
    "    handles = []\n",
    "    for ctype in ctypes:\n",
    "        handle = mpatches.Patch(color=colorDict[ctype], \n",
    "                                ec='k', \n",
    "                                label=ctype)\n",
    "        handles.append(handle)\n",
    "        \n",
    "    plt.scatter(x, y,\n",
    "                c=spotColor,\n",
    "                marker=\"o\",\n",
    "                edgecolor=\"None\",\n",
    "                # lw=0.1,\n",
    "                zorder=2,\n",
    "                alpha=0.7,\n",
    "                s=60)\n",
    "    \n",
    "    # plt.imshow(img.astype(float), \n",
    "    #        interpolation='none',\n",
    "    #        cmap='binary',\n",
    "    #        alpha=0.7,\n",
    "    #        zorder=1)\n",
    "    \n",
    "    xlim = trim[key]['xlim']\n",
    "    ylim = trim[key]['ylim']\n",
    "    plt.xlim(xlim)\n",
    "    plt.ylim(ylim)\n",
    "    plt.yticks([])\n",
    "    plt.xticks([])\n",
    "    \n",
    "    plt.legend(handles=handles, bbox_to_anchor=(1.04, 1.02))\n",
    "    plt.title(key)\n",
    "    plt.show()\n",
    "    \n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af02d6c2-1c8b-479e-ad8f-f0f87156f45c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ctypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6fb22a7-f01d-4631-a509-fa370747d041",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams['figure.dpi'] = 300\n",
    "plt.rcParams['figure.figsize'] = 10, 10\n",
    "plt.rcParams['figure.facecolor'] = 'w'\n",
    "\n",
    "expT = 0 # gene expression threshold\n",
    "\n",
    "\n",
    "for key in keys:\n",
    "    data = adata[key]\n",
    "    # get the image and the scale factors\n",
    "    img, scale_factor = getImage(data)\n",
    "    df = data.to_df()\n",
    "    \n",
    "    # get the image coordinates and scale them\n",
    "    coords = data.obsm['spatial']\n",
    "    x = coords[:, 0]\n",
    "    y = coords[:, 1]\n",
    "    x = x * scale_factor \n",
    "    y = y * scale_factor \n",
    "    \n",
    "    spotType = []\n",
    "    \n",
    "    for idx, row in df.head(10).iterrows():\n",
    "        mask = row > expT\n",
    "        expressedGenes = df.columns[mask]\n",
    "        \n",
    "        ef = gget.enrichr(expressedGenes, database='PanglaoDB_Augmented_2021')\n",
    "        predType = ef.head(1)['path_name'].values[0]\n",
    "        spotType.append(predType)\n",
    "        \n",
    "    \n",
    "    print(pd.Series(spotType).value_counts())\n",
    "\n",
    "   \n",
    "        \n",
    "#     handles = []\n",
    "#     for ctype in ctypes:\n",
    "#         handle = mpatches.Patch(color=colorDict[ctype], \n",
    "#                                 ec='k', \n",
    "#                                 label=ctype)\n",
    "#         handles.append(handle)\n",
    "        \n",
    "#     plt.scatter(x, y,\n",
    "#                 c=spotColor,\n",
    "#                 marker=\"o\",\n",
    "#                 edgecolor=\"None\",\n",
    "#                 # lw=0.1,\n",
    "#                 zorder=2,\n",
    "#                 alpha=0.7,\n",
    "#                 s=60)\n",
    "    \n",
    "#     # plt.imshow(img.astype(float), \n",
    "#     #        interpolation='none',\n",
    "#     #        cmap='binary',\n",
    "#     #        alpha=0.7,\n",
    "#     #        zorder=1)\n",
    "    \n",
    "#     xlim = trim[key]['xlim']\n",
    "#     ylim = trim[key]['ylim']\n",
    "#     plt.xlim(xlim)\n",
    "#     plt.ylim(ylim)\n",
    "#     plt.yticks([])\n",
    "#     plt.xticks([])\n",
    "    \n",
    "#     plt.legend(handles=handles, bbox_to_anchor=(1.04, 1.02))\n",
    "#     plt.title(key)\n",
    "#     plt.show()\n",
    "    \n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a6031ba-3164-4203-9c9e-f6fc62faff65",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
